---
title: "pja_ipmn2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pja_ipmn2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(stgeomx)

library(tidyverse)
library(readxl)
library(writexl)
library(arrow)
library(limma)
library(edgeR)

library(patchwork)
library(ggrepel)
library(ggridges)
library(pheatmap)

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```



```{r input files, include=FALSE}

# master directory
working_dir <- "/Users/mkiyer/Library/CloudStorage/OneDrive-DukeUniversity/research/projects/st_ipmn2"
# input directories
geomx_dir <- file.path(working_dir, "geomx")
# geomx data
geomx_xlsx <- file.path(geomx_dir, "pja_ipmn2_wta_2022-12-02.xlsx")
# sample annotation file
geomx_annot_xlsx <- file.path(geomx_dir, "pja_ipmn2_annotations_2022-12-07.xlsx")

# qc parameters
geomx_negprobe_lod_quantile = 0.9
geomx_min_counts = 50000
geomx_min_auc = 0.65
geomx_aoi_min_frac_expr = 0.2
geomx_gene_min_frac_expr = 0.1

# normalization
norm_method = "qn"

# de parameters
de_padj_cutoff <- 0.05
de_log2fc_cutoff <- 1

# gsea params
gsea_padj_cutoff <- 0.05
gsea_log2fc_cutoff <- 1

plot_dir <- file.path(working_dir, "plots")
if (!dir.exists(plot_dir)) {
  dir.create(plot_dir)
}

```

## Read and process geomx data

```{r geomx data, include=FALSE}

ds <- st_geomx_read(geomx_xlsx)
ds <- st_geomx_process_input(ds, geomx_negprobe_lod_quantile)
ds <- st_geomx_merge_probes(ds)
ds <- st_geomx_rm_empty_aois(ds)

# read and join sample annotation
annot <- read_excel(geomx_annot_xlsx)

ds$samples <- ds$samples %>% 
  inner_join(annot, by="SegmentDisplayName")

table(ds$samples$path)
table(ds$samples$path_specimen)
table(ds$samples$path_aoi)
table(ds$samples$subset)
table(ds$samples$segment)

```


## QC and filtering AOI / Genes

```{r choose filtering thresholds}

summary(ds$samples$num_counts)
summary(ds$samples$frac_expr)

ds <- st_geomx_set_thresholds(ds, 
                              min_counts = geomx_min_counts,
                              min_auc = geomx_min_auc,
                              aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                              gene_min_frac_expr = geomx_gene_min_frac_expr)

table(ds$samples$keep)
table(ds$meta$expressed)

```

## QC: AOI filtering

```{r aoi filtering}

p <- st_geomx_plot_aoi_filter(ds, geomx_min_counts, geomx_min_auc, geomx_aoi_min_frac_expr)
p
ggsave(file.path(plot_dir, "qc_aoi_filter.pdf"), p)

colnames(ds$samples)
ggplot(ds$samples) + 
  geom_point(aes(num_counts, bg_auc, color=bg_cpm)) +
  scale_x_log10() +
  scale_color_viridis_c() +
  facet_wrap(~ segment, scales="free")

# table(ds$samples$keep, ds$samples$segment)

```

## QC: gene filtering cutoff

```{r plot gene fraction detectable above bg}

p <- st_geomx_plot_gene_filter(ds, geomx_gene_min_frac_expr)
p
ggsave(file.path(plot_dir, "qc_gene_filtering_fpr_auc.pdf"), p)

```

## QC: Background correction

```{r background correction}

summary(ds$samples$bg_geomean)
summary(apply(ds$counts, 2, max))

x <- cpm(st_geomx_bgcorrect(ds, method="kdefdr", bw.adjust=5, fdr.min=0.05))
p <- st_geomx_plot_bgcorrect(ds, x)
p
p <- st_geomx_plot_bgcorrect_density(ds, x)
p


```



## Apply filters and normalize

Consolidate technical replicates. When sample has replicates, choose replicate with greater signal-to-noise ratio

```{r filter and normalize}

fds <- list(samples = ds$samples,
            meta = ds$meta,
            counts = ds$counts)

# apply filters
fds <- st_geomx_filter(fds)

# bg correct
#x <- fds$counts
#x <- st_geomx_bgcorrect(fds, method="bgsub")
#x <- st_geomx_bgcorrect(fds, method="norm")
x <- st_geomx_bgcorrect(fds, method="kdefdr", bw.adjust=5, fdr.min=0.05)

# normalize
y_bg_rle <- st_geomx_normalize(fds, x, method="rle")
y_bg_qn <- st_geomx_normalize(fds, x, method="qn")
y_bg_cpm <- st_geomx_normalize(fds, x, method="cpm")

# choose normalized data to work with
normcounts <- y_bg_qn

```



## QC plot: normalization

```{r qc normalization}

# sample annotations to use in plots
s <- select(fds$samples, aoi, slide, keep, path_aoi, path_specimen)
x <- bind_cols(fds$meta, normcounts)
x <- x %>%
  pivot_longer(s$aoi, names_to="aoi", values_to="count") %>%
  inner_join(s, by=c("aoi"="aoi"), suffix=c("_gene", "_aoi"))

y <- x %>%
  group_by(aoi, expressed) %>%
  summarise(slide = unique(slide),
            med = median(count),
            q3 = quantile(count, 0.75)) %>%
  ungroup()

p <- ggplot(x, aes(x=count, y=reorder(aoi, count), fill=slide)) +
  stat_density_ridges(color="#00000066", alpha=0.4, scale=10, rel_min_height=0.001, quantile_lines=TRUE) +
  scale_x_log10() +
  scale_fill_manual(values = pals::cols25()) +
  xlab("Count") +
  ylab("AOIs") +
  labs(fill = "Slide") +
  theme_ridges() +
  theme(axis.text.y = element_blank()) +
  facet_grid(slide ~ ., scales="free_y")
p
f <- file.path(plot_dir, "qc_ridge_plot_norm_aoi.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)


# by aoi path
p <- x %>%
  ggplot(aes(x=count, y=factor(path_aoi), fill=factor(path_aoi))) +
  stat_density_ridges(alpha=0.7, scale=3, rel_min_height=0.001, quantile_lines=TRUE) + 
  scale_color_manual(values = pals::cols25()) + 
  scale_x_log10() +
  theme_ridges() + 
  theme(axis.text = element_text(size=5))
p
f <- file.path(plot_dir, "qc_ridge_plot_norm_path.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)

# by specimen path
p <- x %>%
  ggplot(aes(x=count, y=factor(path_aoi), fill=factor(path_specimen))) +
  stat_density_ridges(alpha=0.7, scale=3, rel_min_height=0.001, quantile_lines=TRUE) + 
  scale_color_manual(values = pals::cols25()) + 
  scale_x_log10() +
  theme_ridges() + 
  theme(axis.text = element_text(size=5))
p
f <- file.path(plot_dir, "qc_ridge_plot_norm_class.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)


```


## Unsupervised clustering / dimensionality reduction

```{r dim reduction}
library(factoextra)
library(umap)

most_cv_genes <- function(x, ntop=500) {
  rv <- apply(x, 1, function(x) sd(x) / mean(x))
  keep <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  return(x[keep,])
}

run_pca <- function(x) {
  pca <- prcomp(t(x))
  pct_var <- round(100 * pca$sdev^2 / sum( pca$sdev^2 ), 2)
  pct_var <- paste(colnames(pca$x), " (", paste(as.character(pct_var), "%", ")", sep=""), sep="")
  return(list(pca = pca, pct_var = pct_var))
}

y <- log2(normcounts)
y <- most_cv_genes(y, 1000)

pca_res <- run_pca(y)
pca_tbl <- bind_cols(fds$samples, as_tibble(pca_res$pca$x))

# pca plots
p <- ggplot(pca_tbl, aes(x=PC1, y=PC2)) +
  theme_minimal() +
  labs(x=pca_res$pct_var[1], y=pca_res$pct_var[2])

# qc factors
p1 <- p + geom_point(aes(color=bg_auc), alpha = 0.8, size=2) +
  scale_color_viridis_c()
p1
ggsave(file.path(plot_dir, "pca_bg_auc.pdf"), plot=p1, device="pdf")

# segment
p1 <- p + geom_point(aes(color=segment), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1
ggsave(file.path(plot_dir, "pca_segment.pdf"), plot=p1, device="pdf")

# slide
p1 <- p + geom_point(aes(color=slide), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1
ggsave(file.path(plot_dir, "pca_path_slide.pdf"), plot=p1, device="pdf")

# path
p1 <- p + geom_point(aes(color=path_aoi, shape=path_specimen), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1
ggsave(file.path(plot_dir, "pca_path_segment.pdf"), plot=p1, device="pdf")


# investigate contributions to pca
res.pca <- prcomp(y, scale = TRUE)
p <- fviz_pca_ind(res.pca, select.ind=list(contrib=20), repel=TRUE, ggrepel.max.overlaps=Inf)
p
ggsave(file.path(plot_dir, "pca_ind.pdf"), p)

p <- fviz_contrib(res.pca, choice = "ind", axes = 1, top=20, max.overlaps=Inf)
p
ggsave(file.path(plot_dir, "pca_contrib.pdf"), p)


# UMAP variable genes
umap_fit <- t(y) %>%
   scale() %>% 
   umap()
umap_df <- umap_fit$layout %>%
   as.data.frame() %>%
   dplyr::rename(umap1 = "V1",
                 umap2 = "V2")
umap_df <- bind_cols(fds$samples, umap_df)


p <- ggplot(umap_df, aes(x=umap1, y=umap2)) +
  theme_minimal()

p1 <- p + geom_point(aes(color=bg_auc), alpha = 0.8, size=2) +
  scale_color_viridis_c()
p1

# segment
p1 <- p + geom_point(aes(color=segment), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1

# slide
p1 <- p + geom_point(aes(color=slide), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1

# segment / path
p1 <- p + geom_point(aes(color=path_aoi, shape=path_specimen), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1



```

## Clustering high variance genes

```{r}
library(pheatmap)

y <- log2(normcounts)
y <- most_cv_genes(y, 1000)

s <- fds$samples %>% mutate(noise_quantile = ntile(desc(bg_auc), 4))

annot_col <- column_to_rownames(s, "aoi") %>% 
  select(slide = slide,
         path_aoi = path_aoi,
         path_specimen = path_specimen,
         noise_quantile = noise_quantile) %>%
  as.data.frame()

p <- pheatmap(y,
         scale="row",
         show_rownames=FALSE,
         show_colnames=FALSE,
         border_color=NA,
         clustering_method = "average",
         clustering_distance_rows = "correlation",
         clustering_distance_cols = "correlation",
         breaks = seq(-3, 3, 0.05),
         color = colorRampPalette(c("purple3", "black", "yellow2"))(120),
         annotation_col = annot_col)

f <- file.path(plot_dir, "heatmap_cluster_highcv_genes.pdf")
save_pheatmap_pdf(p, filename=f, width=12, height=8)

```

## Differential expression analysis

```{r de analysis setup}

limma_rename_result_cols <- function(res) {
  as_tibble(res, rownames="gene") %>%
    select(gene,
           log2fc=logFC,
           avgexpr=AveExpr,
           pval=P.Value,
           padj=adj.P.Val)
}

de_format_results <- function(res, analysis, method, 
                              padj_cutoff, log2fc_cutoff) {
  mutate(res, 
         de = case_when(padj > padj_cutoff ~ "no",
                        log2fc < -log2fc_cutoff ~ "dn",
                        log2fc > log2fc_cutoff ~ "up",
                        TRUE ~ "no"),
         analysis = analysis,
         method = method)
}

run_limma_trend <- function(y, design, contrasts) {
  fit <- lmFit(y, design)
  fit <- contrasts.fit(fit, contrasts)
  fit <- eBayes(fit, trend=TRUE)
  return(fit)
}

process_limma <- function(fit, contrasts, method, 
                          padj_cutoff, log2fc_cutoff) {
  x <- NULL
  for (coef in colnames(contrasts)) {
    res <- topTable(fit, coef=coef, number=Inf, sort.by="none")
    res <- limma_rename_result_cols(res)
    res <- de_format_results(res, 
                             analysis=coef, 
                             method=method,
                             padj_cutoff=padj_cutoff, 
                             log2fc_cutoff=log2fc_cutoff)
    x <- bind_rows(x, res)
  }
  return(x)
}


plot_de_volcano <- function(res) {
  # res <- res %>% 
  #   mutate(de = case_when(padj > 0.05 ~ "no",
  #                         log2fc < 0 ~ "dn",
  #                         log2fc > 0 ~ "up"))
  p <- ggplot(res, aes(x=log2fc, y=-log10(padj), color=de, size=de)) +
    geom_point(alpha=0.7) +
#    geom_text_repel(data=subset(res, de != "no"), color="black", size=3, aes(label=gene), max.overlaps=Inf) +
    geom_text_repel(data=subset(res, de != "no"), color="black", size=3, aes(label=gene)) +
    scale_color_manual(values=c("no"="grey", "dn"="blue", "up"="red")) +
    scale_size_manual(values=c("no"=1, "dn"=2, "up"=2)) +
    theme_minimal() +
    theme(legend.position="bottom") + 
    theme(axis.line = element_line(color = "black"))
    labs(x="log2fc", y="-log10(padj)")
  return(p)
}


```


## DE analysis: epithelial compartment, specimens with invasive carcinoma

```{r between de analysis pathology}

table(fds$samples$path_aoi, fds$samples$slide)

s <- fds$samples %>% filter(path_specimen == "inv", segment == "panck")
s$path <- factor(tolower(s$path_aoi), levels=c("nl", "lgd", "hgd", "inv"))
y <- log2(normcounts)[,s$aoi]

# de analysis
de <- NULL
method <- "limma_trend"

design <- stats::model.matrix(~0 + slide + path, data=s)
colnames(design)

table(s$slide, s$path)



colnames(design) <- levels(s$path)
s$path
contrasts <- limma::makeContrasts(lgd_vs_nl = panck_lgd - panck_nmld,
                                  hgd_vs_nl = panck_hgd - panck_nmld,
                                  inv_vs_nl = panck_inv - panck_nmld,
                                  hgd_vs_lgd = panck_hgd - panck_lgd,
                                  inv_vs_lgd = panck_inv - panck_lgd,
                                  inv_vs_hgd = panck_inv - panck_hgd,
                                  invhgd_vs_lgdnl = (panck_inv + panck_hgd)/2 - (panck_lgd + panck_nmld)/2,
                                  levels=design)
fit <- run_limma_trend(y, design, contrasts)
res <- process_limma(fit, contrasts, method, de_padj_cutoff, de_log2fc_cutoff) 
de <- bind_rows(de, res)


model.matrix(~slide, data=s)

# 
# # design
# on_icb_response <- s$time == "t1" & s$response == "yes"
# on_icb_no_response <- s$time == "t1" & s$response == "no"
# design <- model.matrix(~patient, data=s)
# design_response <- cbind(design, on_icb_no_response, on_icb_response)
# 
# y <- log2(g_qn_cpm + log2_prior_count)
# y <- get_filtered_subset(norm_expr_mat[, s$library], 
#                          min.expr = de_min_log2_cpm, 
#                          min.prop = de_min_prop)
# print(nrow(y))
# res <- run_limma_trend(y, design_response, "on_icb_response", 
#                        de_padj_cutoff, de_log2fc_cutoff,
#                        analysis, method)
# de <- bind_rows(de, res)

write_xlsx(de, file.path(working_dir, "de_path.xlsx"))



res %>% filter(gene == "LAMB3")
res %>% filter(analysis == "hgd_vs_lgd") %>% arrange(padj)
table(res$analysis, (res$de == "up") & (res$padj < 1e-3))


write_xlsx(de)

plot_de_volcano(res) +
  facet_wrap(~ analysis)
  
p


  labs(title = "Acral vs. Non-Acral")
p
ggsave(file.path(plot_dir, "de_volcano_acral.pdf"), p)


# Survival 3yr
analysis <- "os_3yr"
s$os_3yr <- factor(ifelse(s$os_quartile < 1, "under3yrs", "over3yrs"), levels=c("over3yrs", "under3yrs"))

design <- stats::model.matrix(~ 0 + os_3yr, data=s)
colnames(design) <- levels(s$os_3yr)
contrasts <- limma::makeContrasts(os3yr = under3yrs - over3yrs,
                                  levels=design)

fit <- run_limma_trend(y, design, contrasts)
res <- process_limma(fit, contrasts, method, de_padj_cutoff, de_log2fc_cutoff) 
de <- bind_rows(de, res)
p <- plot_de_volcano(res) + ggtitle("OS <3yr vs >3yr")
p


# Survival 5yr
analysis <- "os_5yr"
s$os_5yr <- factor(ifelse(s$os_quartile < 2, "under5yrs", "over5yrs"), levels=c("over5yrs", "under5yrs"))

design <- stats::model.matrix(~ 0 + os_5yr, data=s)
colnames(design) <- levels(s$os_5yr)
contrasts <- limma::makeContrasts(os5yr = under5yrs - over5yrs,
                                  levels=design)

fit <- run_limma_trend(y, design, contrasts)
res <- process_limma(fit, contrasts, method, de_padj_cutoff, de_log2fc_cutoff) 
de <- bind_rows(de, res)
p <- plot_de_volcano(res) + ggtitle("OS <5yr vs >5yr")
p



# Survival 9 yr
analysis <- "os_9yr"
s$os_9yr <- factor(ifelse(s$os_quartile < 3, "under9yrs", "over9yrs"), levels=c("over9yrs", "under9yrs"))

design <- stats::model.matrix(~ 0 + os_9yr, data=s)
colnames(design) <- levels(s$os_9yr)
contrasts <- limma::makeContrasts(os9yr = under9yrs - over9yrs,
                                  levels=design)
fit <- run_limma_trend(y, design, contrasts)
res <- process_limma(fit, contrasts, method, de_padj_cutoff, de_log2fc_cutoff) 
de <- bind_rows(de, res)
p <- plot_de_volcano(res) + ggtitle("OS <9yr vs >9yr")
p
ggsave(file.path(plot_dir, "de_volcano_os9yr.pdf"), p)



# Survival q0 vs q3
table(s$os_quartile)

s$os_quartile <- factor(s$os_quartile)
design <- stats::model.matrix(~ 0 + os_quartile, data=s)
contrasts <- limma::makeContrasts(os_q0_vs_q3 = os_quartile0 - os_quartile3,
                                  levels=design)
fit <- run_limma_trend(y, design, contrasts)
res <- process_limma(fit, contrasts, method, de_padj_cutoff, de_log2fc_cutoff) 

de <- bind_rows(de, res)
table(res$analysis)
p <- plot_de_volcano(res) + ggtitle("OS <3yr vs >9yr")
p
ggsave(file.path(plot_dir, "de_volcano_os_q0_vs_q3.pdf"), p)


de %>% filter(analysis == "acral") %>% arrange(padj)
de %>% filter(padj < 0.05)

de %>% filter(gene == "EPHA4")
# de %>% arrange(desc(avgexpr), desc(log2fc)) %>% distinct(gene, .keep_all=TRUE)
# de %>% filter(gene == "LZTR1")

# write gene expression data
#write_xlsx(de_merged,
#           file.path(working_dir, 'de_results.xlsx'))

```

## plot individual genes

```{r plot genes}

plot_gene <- function(s, x, gene, aes_x, aes_color) {
  exprs <- unlist(x[gene, ])
  x <- s %>% add_column(gene = exprs)
  p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
    geom_boxplot(outlier.shape = NA, width=0.5) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
    scale_fill_manual(values=pals::cols25()) +
    ggtitle(paste0("Gene: ", gene)) +
    theme_minimal() + 
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
    theme(axis.line = element_line(color = "black"))
  return(p)
}


de %>% filter(gene == "PMM1")
p <- plot_gene(s, y, "PMM1", acral, acral)
p

p <- plot_gene(s, y, "PI15", os_quartile, os_quartile)
p
ggsave(file.path(plot_dir, "gene_afap1l2_os.pdf"), p)


plot_gene(s, y, "ATP5MC3", os_quartile, os_quartile)
de %>% filter(padj < 0.05)


```







## BG correction exploration

```{r bg correction exploration}

bg <- ds$meta$bg
x <- ds$counts

sort(ds$samples$bg_auc)[1:20]
sort(ds$samples$bg_auc, decreasing=TRUE)[1:10]
s <- "s08_r010_panck"
xs <- unlist(x[,s])

bw.adjust <- 5

x <- log2(xs)
xmin <- min(x)
xmax <- max(x)

n <- 10000
bw <- bw.nrd0(x) * bw.adjust
eps <- 1e-10

dbg <- density(x[bg], bw=bw, from=xmin, to=xmax, n=n)

?quantile


#d <- density(x[!noise], bw=bw, from=xmin, to=xmax, n=n)

dcdf <- cumsum(d$y) / sum(d$y)
dbgcdf <- cumsum(dbg$y) / sum(dbg$y)
dbgcdf <- pmax(dbgcdf, dcdf)

dfdr <- (1-dbgcdf) / (1-dcdf+eps)
dfdr <- pmin(1, dfdr + fdr.min)
scaled_fdr <- xmin + (xmax - xmin) * cumsum(dfdr) / n
scaled_fdr <- pmin(d$x, scaled_fdr)
xnoise <- approx(d$x, scaled_fdr, xout=x)$y
xnoise <- pmin(xnoise, x)
y <- 2^x - 2^xnoise + 1

#dfdr <- (1-dbgcdf) / (1-dcdf)
#dfdr[which(is.nan(dfdr))] <- min(dfdr[which(!is.nan(dfdr))])
#dfdr <- pmin(1, dfdr + fdr.min)



res <- cluster_noise(xs, bg)

tbl <- bind_cols(x=res$x,
                 bg=res$bg_dist,
                 hi=res$hi_dist,
                 tot=res$bg_dist + res$hi_dist)

ggplot(tbl) +
  geom_point(aes(x,bg), color="red", alpha=0.5) +
  geom_point(aes(x,hi), color="blue", alpha=0.5) +
  geom_point(aes(x,tot)) +
  geom_vline(xintercept=res$bg_center, linetype="dashed", color="red") +
  geom_vline(xintercept=res$hi_center, linetype="dashed", color="blue") +
  geom_vline(xintercept=res$cutoff, linetype="dashed")


res <- bgcorrect_kdefdr2(xs, bg, bw.adjust=1, fdr.min=0.05)$y


for (s in ds$samples$aoi) {
  print(s)
  xs <- unlist(x[,s])
  res <- bgcorrect_kdefdr2(xs, bg, bw.adjust=1, fdr.min=0.05)$y
}



```
