---
title: "pja_ipmn_wta"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pja_ipmn2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(stgeomx)

library(tidyverse)
library(readxl)
library(writexl)
library(limma)
library(factoextra)
library(umap)
library(fgsea)
library(msigdbr)

library(patchwork)
library(ggrepel)
library(ggridges)
library(pheatmap)

```

## Input files and parameter settings

```{r input files, include=FALSE}

# master directory
working_dir <- "/Users/mkiyer/Library/CloudStorage/OneDrive-DukeUniversity/research/projects/st_ipmn2"
data_dir <- file.path(working_dir, "ext")

# input directories
geomx_dir <- file.path(working_dir, "geomx")

# WTA PKC file
wta_pkc_file <- file.path(geomx_dir, "Hs_R_NGS_WTA_v1.0.pkc")

# Peter Allen Lab IPMN WTA dataset
pja_ipmn_geomx_xlsx <- file.path(geomx_dir, "pja_ipmn2_wta_2022-12-02.xlsx")
pja_ipmn_geomx_annot_xlsx <- file.path(geomx_dir, "pja_ipmn2_annotations_2023-04-13.xlsx")

# Spatial Organ Atlas dataset
soa_panc_geomx_xlsx <- file.path(geomx_dir, "soa_panc_2022-02-21.xlsx")
soa_panc_geomx_annot_xlsx <- file.path(geomx_dir, "soa_panc_annotations_2023-04-11.xlsx")

# GSE226829 (donor pancreata)
donor_geomx_probes_xlsx <- file.path(geomx_dir, "GSE226829_PreQC_probes.xlsx")
donor_geomx_counts_xlsx <- file.path(geomx_dir, "GSE226829_PreQC_raw_counts.xlsx")
donor_geomx_samples_xlsx <- file.path(geomx_dir, "GSE226829_PreQC_segments_annotation_edited.xlsx")
donor_geomx_dcc_tar <- file.path(geomx_dir, "GSE226829_RAW.tar")

# external resource files
hpa_path_file <- file.path(data_dir, "hpa_pathology.tsv")
cptac_panc_file <- file.path(data_dir, "cptac_panc_supp_mmc3.xlsx")

# qc parameters
geomx_negprobe_lod_quantile = 0.9
geomx_min_counts = 50000
geomx_min_auc = 0.65
geomx_aoi_min_frac_expr = 0.2
geomx_gene_min_frac_expr = 0.1

# normalization
norm_method = "qn"
bgcorrect_method = "bgsub"
bgcorrect_bw_adjust = 2
bgcorrect_bg_quant = 0.5
norm_method = "qn"

# de parameters
de_padj_cutoff <- 0.05
de_log2fc_cutoff <- 1

# gsea params
gsea_padj_cutoff <- 0.05
gsea_log2fc_cutoff <- 1

plot_dir <- file.path(working_dir, "plots")
if (!dir.exists(plot_dir)) {
  dir.create(plot_dir)
}

gsea_plot_dir <- file.path(working_dir, "gsea_plots")
if (!dir.exists(gsea_plot_dir)) {
  dir.create(gsea_plot_dir)
}

```

## Helper functions

```{r helper functions, include=FALSE}

get_color_scales <- function(patients) {
  color_scales = list(
    path_aoi = c(nl="#0055ff", lgd="#00FFFF", hgd="#ffdd00", inv="#ff0000"),
    path_specimen = c(lgd="#0000cc", inv="#cc0000"),
    path_patient = c(lgd_nl="#4B0082", lgd_lgd="#d6bfff", inv_nl="#0055ff", inv_lgd="#00FFFF", inv_hgd="#ffdd00", "inv_inv"="#ff0000"),
    de_cptac_rna = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"), 
    de_cptac_prot = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF")
  )
  x <- pals::cols25(length(unique(patients)))
  color_scales$patient = setNames(x, sort(unique(patients)))
  return(color_scales)
}


# heatmap functions
plot_pheatmap <- function(m, 
                          annot_col, 
                          annot_row, 
                          annot_colors) {
  pheatmap(m,
           scale = "none",
           color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
           breaks = seq(-6, 6, length.out=51),
           annotation_col = annot_col,
           annotation_row = annot_row,
           annotation_colors = annot_colors,
           border_color = NA,
           cluster_rows = TRUE,
           cluster_cols = TRUE,
           clustering_distance_rows = "euclidean",
           clustering_distance_cols = "euclidean",
           clustering_method = "ward.D2",
           fontsize_row = 6,
           fontsize_col = 6)
}

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

get_row_annot <- function(genes, gene_annot) {
  gene_annot %>%
    filter(gene_symbol %in% genes) %>%
    select(gene_symbol, prognostic, de_cptac_rna, de_cptac_prot, grade_subtype, hist_subtype) %>%
    column_to_rownames("gene_symbol") %>%
    as.data.frame()
}



```

## Read and process IPMN WTA geomx data

```{r ipmn wta geomx data, include=FALSE}

# read data
ds <- st_geomx_read_xlsx(pja_ipmn_geomx_xlsx)
# read and join sample annotation
annot <- read_excel(pja_ipmn_geomx_annot_xlsx)
ds$samples <- ds$samples %>% 
  inner_join(annot, by="SegmentDisplayName")

# subset epithelial segment
s <- ds$samples %>% filter(segment == "panck")
ds <- list(samples = s,
           meta = ds$meta,
           counts = ds$counts[, s$aoi])
# name dataset
ipmn_ds <- ds

```

## Read and process Spatial Organ Atlas pancreas geomx data

```{r soa panc geomx data, include=FALSE}

# read data
ds <- st_geomx_read_xlsx(soa_panc_geomx_xlsx)
# read and join sample annotation
annot <- read_excel(soa_panc_geomx_annot_xlsx)
ds$samples <- ds$samples %>% 
  inner_join(annot, by="SegmentDisplayName")

# subset epithelial segment
s <- ds$samples %>% filter(segment == "panck+")
ds <- list(samples = s,
           meta = ds$meta,
           counts = ds$counts[, s$aoi])
# name dataset
soa_panc_ds <- ds

```

## Read Donor Pancreas WTA Study

```{r read wta pkc, include=FALSE}
library(GeomxTools)

st_geomx_read_pkc <- function(pkc_file) {
  pkc <- yaml::read_yaml(pkc_file)
  probes <- tibble(target = pkc$Targets)
  probes <- probes %>% tidyr::unnest_wider(target) %>%
    dplyr::rename(TargetName = DisplayName) %>%
    tidyr::unnest_longer(Probes) %>%
    tidyr::unnest_wider(Probes) %>%
    dplyr::rename(ProbeDisplayName = DisplayName) %>%
    dplyr::mutate(bg = startsWith(CodeClass, "Negative"))
  return(probes)
}

st_geomx_read_dcc <- function(dcc_tar_file, dcc_dir) {
  # extract dcc tar file
  untar(dcc_tar_file, exdir=dcc_dir)
  # read dcc files
  dcc_files <- list.files(dcc_dir, full.names=TRUE)
  dcc_data <- sapply(dcc_files, GeomxTools::readDccFile, simplify=FALSE)
  dcc_data <- tibble(dcc = dcc_data) %>%
    unnest_wider(dcc, simplify=FALSE)
  
  samples <- dcc_data %>% 
    select(Header, Scan_Attributes, NGS_Processing_Attributes) %>%
    unnest(cols = c(Header, Scan_Attributes, NGS_Processing_Attributes))
  
  counts <- dcc_data %>%
    select(Scan_Attributes, Code_Summary) %>% 
    hoist(Scan_Attributes, SampleID="SampleID") %>% 
    select(-Scan_Attributes) %>%
    unnest(Code_Summary) %>% 
    pivot_wider(names_from=SampleID, values_from=Count, values_fill=1)
  
  # TODO: not sure whether to leave values at zero or 1
  # TODO: could add one to all counts
  # TODO: needs to be consistent with the way DSP server processes
  
  return(list(samples=samples,
              counts=counts))
}

# read pkc file
probes <- st_geomx_read_pkc(wta_pkc_file)

# extract dcc data from tar file
dcc_dir <- file.path(working_dir, "dcc")
ds <- st_geomx_read_dcc(donor_geomx_dcc_tar, dcc_dir)

# read and join sample annotation
samples <- read_excel(donor_geomx_samples_xlsx) %>%
  inner_join(ds$samples, by=c("SampleID"="SampleID"))

# join probes with counts
counts <- left_join(probes, ds$counts, by="RTS_ID")

# ensure all samples found in counts
inds <- match(pull(samples, SampleID), colnames(counts))
stopifnot(all(!is.na(inds)))

# break counts into metadata and count data
meta <- select(counts, rts=RTS_ID, probe=ProbeDisplayName, gene=TargetName, bg)

# select valid samples
counts <- select(counts, all_of(inds))
# ensure samples and count tables are aligned
stopifnot(all(pull(samples, SampleID) == colnames(counts)))

# create new identifiers for slide, roi, segment, aoi
slide_keys <- pull(samples, SlideName)
samples$slide <- sprintf("s%02d", match(slide_keys, unique(slide_keys)))
samples <- samples %>% mutate(
  roi = sprintf("r%s", ROILabel),
  segment = tolower(gsub(" ", "", SegmentLabel, fixed = TRUE)),
  aoi = sprintf("%s_%s_%s", slide, roi, segment)
)
stopifnot(n_distinct(samples$aoi) == nrow(samples))

# relabel columns of count matrix with new aoi id
colnames(counts) <- samples$aoi
stopifnot(all(samples$aoi == colnames(counts)))

ds <- list(samples=samples,
           meta=meta,
           counts=counts)

# subset epithelial segment
s <- ds$samples %>% filter(segment == "panck")
ds <- list(samples = s,
           meta = ds$meta,
           counts = ds$counts[, s$aoi])
# name dataset
donor_ds <- ds

```

## Merge GeoMx datasets

```{r merge geomx datasets, include=FALSE}

add_study_field <- function(ds, study) {
  s <- ds$samples
  s$study <- study
  s <- s %>% mutate(patient = sprintf("%s_%s", study, patient),
                    slide = sprintf("%s_%s", study, slide),
                    roi = sprintf("%s_%s", study, roi),
                    aoi = sprintf("%s_%s", study, aoi))
  colnames(ds$counts) <- s$aoi
  return(list(samples = s, meta = ds$meta, counts = ds$counts))
}

merge_ds <- function(a, b) {
  # merge sample tables
  sample_cols <- intersect(colnames(a$samples), colnames(b$samples))
  s <- bind_rows(select(a$samples, all_of(sample_cols)),
                 select(b$samples, all_of(sample_cols)))
  # merge count data
  a_counts <- bind_cols(select(a$meta, probe, gene, bg), a$counts)
  b_counts <- bind_cols(select(b$meta, probe, gene, bg), b$counts)
  counts <- inner_join(a_counts, b_counts, by=c("probe", "gene", "bg"),
                       relationship="one-to-one")
  # counts <- inner_join(a_counts, b_counts, by=c("probe", "gene", "bg"),
  #                      suffix=c(paste0("_", a_study), paste0("_", b_study)),
  #                      relationship="one-to-one")
  
  # break into metadata and count data
  meta <- select(counts, -all_of(s$aoi))
  counts <- select(counts, all_of(s$aoi))
  # fraction expressed 
  # meta <- mutate(meta, frac_expr = pmax(!!rlang::sym(paste0("frac_expr_", a_study)),
  #                                       !!rlang::sym(paste0("frac_expr_", b_study))))
  ds <- list(samples=s, meta=meta, counts=counts)
  return(ds)
}

# add study field
ipmn_ds <- add_study_field(ipmn_ds, "ipmn")
soa_panc_ds <- add_study_field(soa_panc_ds, "soa_panc")
donor_ds <- 
  
# merge datasets
ds <- merge_ds(add_study_field(ipmn_ds, "ipmn"), 
               add_study_field(soa_panc_ds, "soa_panc"))
ds <- merge_ds(ds, add_study_field(donor_ds, "donor"))


# process
ds <- st_geomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- st_geomx_merge_probes(ds)
ds <- st_geomx_rm_empty_aois(ds)

```

## QC and filtering AOI / Genes

```{r choose filtering thresholds}

summary(ds$samples$num_counts)
summary(ds$samples$frac_expr)


p <- ggplot(ds$samples, aes(patient, frac_expr, fill=study)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1))
p




ds <- st_geomx_set_thresholds(ds, 
                              min_counts = geomx_min_counts,
                              min_auc = geomx_min_auc,
                              aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                              gene_min_frac_expr = geomx_gene_min_frac_expr)

# filter AOIs
p <- st_geomx_plot_aoi_filter(ds, geomx_min_counts, geomx_min_auc, geomx_aoi_min_frac_expr)
p
ggsave(file.path(plot_dir, "qc_aoi_filter.pdf"), p)

# filter genes
p <- st_geomx_plot_gene_filter(ds, geomx_gene_min_frac_expr)
p
ggsave(file.path(plot_dir, "qc_gene_filtering_fpr_auc.pdf"), p)

# apply filters
fds <- list(samples = ds$samples,
            meta = ds$meta,
            counts = ds$counts)
fds <- st_geomx_filter(fds)


nrow(ds$samples)
table(ds$meta$keep)
nrow(fds$samples)

color_scales <- get_color_scales(fds$samples$patient)

```
## Normalization

```{r norm}

method <- "bgsub_qn"

bgsub_ceiling <- function(x, offset=0, min.count=1) {
  x <- ceiling(sweep(x, 2, offset, FUN="-"))
  x <- apply(x, 2, pmax, min.count)
  x
}

tbqnorm <- function(x) {
  # convert to relative counts (counts-per-million)
  xcpm <- apply(x, 2, function(x) { 1e6 * x / sum(x) })
  # rank genes across samples by total relative expression
  global_ranks <- rank(rowSums(xcpm), ties.method="random")
  # use correlation to compute sample-sample similarity matrix
  xcor <- cor(xcpm, method="spearman")
  # exclude samples with negative correlation
  xcor[xcor < 0] <- NA
  # ordered list of nearest neighbors for each sample
  nn_list <- apply(xcor, 2, order, na.last=NA, decreasing=TRUE)
  # starting ranks (with ties present)
  xrank_init <- as.data.frame(apply(x, 2, rank, ties.method="min"))
  #xrank_init <- reframe(as_tibble(x), across(everything(), min_rank))
  # reordering procedure to break ties for each sample
  xrank <- matrix(nrow=nrow(x), ncol=ncol(x), dimnames=list(NULL, colnames(x)))
  for(i in 1:ncol(x)) {
    nn_ordered_cols <- xrank_init[, nn_list[,i]]
    nn_ordered_cols <- bind_cols(nn_ordered_cols, tiebrk_ranks=global_ranks)
    xrank[,i] <- do.call(order, unname(nn_ordered_cols))
  }
  # convert ordering to ranks
  xrank <- apply(xrank, 2, order)

  # compute quantiles
  xquantiles <- apply(xcpm, 2, sort)
  xquantiles <- apply(xquantiles, 1, geomean)
  # renormalize quantiles such that columns sum to one million (e.g. cpm)
  xquantiles <- 1e6 * (xquantiles / sum(xquantiles))
  # apply quantile normalization procedure using new ranks
  xnorm <- apply(xrank, 2, function(x) { xquantiles[x] })
  return(xnorm)
}

# normalize genes
x <- bgsub_ceiling(fds$counts, fds$samples$bg_geomean, min.count=0)
x <- tbqnorm(x)
rownames(x) <- fds$meta$gene

# filter genes that are zero
keep_genes <- rowSums(x) > 0
x <- x[keep_genes, ]
fds <- list(
  samples = fds$samples,
  meta = filter(fds$meta, keep_genes),
  counts = fds$counts[keep_genes, ],
  bgmeta = fds$bgmeta,
  bgcounts = fds$bgcounts
)

norm_cpm <- x
log2_norm_cpm <- log2(x+1)

```


## Read external datasets

```{r read external data}

get_top_gene_sets <- function(tbl, log2fc_cutoff, padj_cutoff) {
  ntopgenes <- c(20, 50, 100, 200, 500)
  gs <- list()
  for (a in unique(tbl$analysis)) {
    up_genes <- tbl %>% filter(
      analysis == a,
      log2fc >= log2fc_cutoff,
      fdr < padj_cutoff
    )
    dn_genes <- tbl %>% filter(
      analysis == a,
      log2fc <= -log2fc_cutoff,
      fdr < padj_cutoff
    )
    gs_name <- paste0(toupper(a), "_ALL_UP")
    gs[[gs_name]] <- pull(up_genes, gene_symbol)
    gs_name <- paste0(toupper(a), "_ALL_DN")
    gs[[gs_name]] <- pull(dn_genes, gene_symbol)

    for (n in ntopgenes) {
      x <- up_genes %>% slice_max(log2fc, n=n) %>% pull(gene_symbol)
      if (length(x) < n) { next }
      gs_name <- paste0(toupper(a), "_TOP", n, "_UP")
      gs[[gs_name]] <- x
      
      x <- dn_genes %>% slice_min(log2fc, n=n) %>% pull(gene_symbol)
      if (length(x) < n) { next }
      gs_name <- paste0(toupper(a), "_TOP", n, "_DN")
      gs[[gs_name]] <- x 
    }
  }
  return(gs)
}

read_cptac_pdac_data <- function(filename, log2fc_cutoff, padj_cutoff) {
  rna <- read_excel(filename, sheet="Diff_exp_RNA")
  rna <- rna %>%
    dplyr::rename(gene_symbol = GeneSymbol,
                  log2fc = MedianLog2FoldChange,
                  pval = `p value`,
                  fdr = FDR) %>%
    dplyr::mutate(analysis = "cptac_rna", 
                  de = case_when(fdr > padj_cutoff ~ "no",
                                 abs(log2fc) <= log2fc_cutoff ~ "no",
                                 log2fc < -log2fc_cutoff ~ "dn",
                                 log2fc > log2fc_cutoff ~ "up",
                                 TRUE ~ "no"))
  
  # GeneSymbol	MedianLog2FoldChange	p value	FDR	
  # MedianLog2FoldChange.duct	p value.duct	FDR.duct	
  # Expression Coefficient	p value of adjusted expression	FDR of adjusted expression
  # add additional data?
  prot <- read_excel(cptac_panc_file, sheet="Diff_exp_prot")
  prot <- prot %>%
    dplyr::rename(
      gene_symbol = GeneSymbol,
      log2fc = MedianLog2FoldChange,
      pval = `p value`,
      fdr = FDR
    ) %>%
    select(gene_symbol, log2fc, pval, fdr) %>%
    mutate(analysis = "cptac_prot",
           de = case_when(fdr > padj_cutoff ~ "no",
                          abs(log2fc) <= log2fc_cutoff ~ "no",
                          log2fc < -log2fc_cutoff ~ "dn",
                          log2fc > log2fc_cutoff ~ "up",
                          TRUE ~ "no"))
  return(bind_rows(rna, prot))
}


read_hpa_data <- function(hpa_path_file) {
  x <- read.table(hpa_path_file, header=TRUE, sep="\t")
  colnames(x) <- c("ens_gene_id", "gene_symbol", "cancer_type",
                   "ihc_high", "ihc_medium", "ihc_low", "ihc_not_detected",
                   "prog_fav", "unprog_fav", "prog_unfav", "unprog_unfav")
  x <- as_tibble(x) %>%
    distinct(gene_symbol, cancer_type, .keep_all=TRUE) %>%
    mutate(cancer_type = str_replace(cancer_type, " ", "_")) %>%
    mutate(prognostic = case_when(!is.na(prog_fav) ~ "fav",
                                  !is.na(prog_unfav) ~ "unfav",
                                  TRUE ~ "no")) %>%
    select(ens_gene_id, gene_symbol, cancer_type, prognostic, prog_fav, prog_unfav)
  return(x)
}

get_hpa_gene_sets <- function(hpa) {
  gs = list()
  for (catype in unique(hpa$cancer_type)) {
    unfav <- filter(hpa, cancer_type == catype, prognostic == "unfav")  %>% select(gene_symbol) %>% pull()
    fav <- filter(hpa, cancer_type == catype, prognostic == "fav")  %>% select(gene_symbol) %>% pull()
    unfav_name <- paste0("HPA_", toupper(catype), "_UNFAV_PROGNOSIS")
    fav_name <- paste0("HPA_",  toupper(catype), "_FAV_PROGNOSIS")
    x <- setNames(list(unfav, fav), c(unfav_name, fav_name))
    gs <- append(gs, x)
  }
  gs
}

# read cptac data
cptac_de_data <- read_cptac_pdac_data(cptac_panc_file, gsea_log2fc_cutoff, gsea_padj_cutoff)
cptac_de_data <- filter(cptac_de_data, gene_symbol %in% fds$meta$gene)
# pivot to merge protein/rna data
cptac_de_merged <- cptac_de_data %>%
  pivot_wider(names_from = analysis, values_from = c(log2fc, pval, fdr, de))
# get gene sets
gs_cptac <- get_top_gene_sets(cptac_de_data, gsea_log2fc_cutoff, gsea_padj_cutoff)

# read hpa data
hpa <- read_hpa_data(hpa_path_file)
hpa <- filter(hpa, gene_symbol %in% fds$meta$gene)
hpa_pdac <- filter(hpa, cancer_type == "pancreatic_cancer")
# get gene sets
gs_hpa <- get_hpa_gene_sets(hpa)
gs_hpa_panc <- gs_hpa[c("HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS", 
                        "HPA_PANCREATIC_CANCER_FAV_PROGNOSIS")]

# read msigdb data
msigdb_gene_sets = msigdbr(species = "Homo sapiens")
msigdb_gene_sets <- msigdb_gene_sets %>% 
  filter(gene_symbol %in% fds$meta$gene) %>%
  distinct(gs_name, gene_symbol, .keep_all = TRUE)

# filter gruetzmann microarray dataset
gruetzmann_gs_names <- c("GRUETZMANN_PANCREATIC_CANCER_UP",
                         "GRUETZMANN_PANCREATIC_CANCER_DN")
gruetzmann_gs <- msigdb_gene_sets %>%
  filter(gs_name %in% gruetzmann_gs_names)
gruetzmann_gs <- split(x = gruetzmann_gs$gene_symbol,
                       f = gruetzmann_gs$gs_name)

# PDAC enrichment gene sets
# gs_pdac <- gs_cptac
gs_pdac <- gruetzmann_gs
gs_pdac <- append(gs_pdac, gs_hpa_panc)
gs_pdac <- append(gs_pdac, gs_cptac)
lapply(gs_pdac, length)

# discovery analysis across msigdb gene sets
gs_hallmark <- filter(msigdb_gene_sets, gs_cat == "H")
gs_hallmark <- split(x = gs_hallmark$gene_symbol, f = gs_hallmark$gs_name)
gs_gobp <- filter(msigdb_gene_sets, gs_cat == "C5", gs_subcat == "GO:BP")
gs_gobp <- split(x = gs_gobp$gene_symbol, f = gs_gobp$gs_name)

```


## PCA

```{r analysis}

most_cv_genes <- function(x, ntop=500) {
  rv <- apply(x, 1, function(x) sd(x) / mean(x))
  keep <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  return(x[keep,])
}

most_var_genes <- function(x, ntop=500, span=0.50) {
  tbl <- bind_cols(
    u = rowMeans(x),
    v = apply(x, 1, function(x) var(x))
  )
  lovar <- loess(v ~ u, data=tbl, span=0.50)
  tbl$lovar <- predict(lovar)
  tbl$vadj <- abs(tbl$v / tbl$lovar)
  keep <- order(tbl$vadj, decreasing=TRUE)[seq_len(min(ntop, nrow(tbl)))]
  return(x[keep,])
}


run_pca <- function(x) {
  pca <- prcomp(t(x))
  pct_var <- round(100 * pca$sdev^2 / sum( pca$sdev^2 ), 2)
  pct_var <- paste(colnames(pca$x), " (", paste(as.character(pct_var), "%", ")", sep=""), sep="")
  return(list(pca = pca, pct_var = pct_var))
}

pca_plot <- function(pca_res, pca_tbl, color_by, shape_by) {
  p <- ggplot(pca_tbl, aes(x=PC1, y=PC2, color={{color_by}}, shape={{shape_by}})) +
    geom_point(alpha = 0.8, size=2) +
    theme_bw() +
    labs(x=pca_res$pct_var[1], y=pca_res$pct_var[2])
  return(p)
}

pca_bg_fit <- function(x, response, npcs=5) {
  # pca
  pca_res <- prcomp(t(x), center=TRUE, scale=TRUE)
  pct_var <- pca_res$sdev^2 / sum(pca_res$sdev^2)
  tot_rsq <- 0
  for (i in 1:npcs) {
    tbl <- bind_cols(y = response, x = pca_res$x[,i])
    model <- lm(y ~ x, data=tbl)
    msum <- summary(model)
    rsq <- msum$r.squared * pct_var[i]
    tot_rsq <- tot_rsq + rsq
  }
  return(tot_rsq)
}

y <- log2_norm_cpm
#pca_res <- run_pca(most_var_genes(y, ntop=500))
pca_res <- run_pca(y)
pca_tbl <- bind_cols(fds$samples, as_tibble(pca_res$pca$x))

width <- 5
height <- 4

# bg_ks_dist
p1 <- pca_plot(pca_res, pca_tbl, bg_ks_dist, path_specimen) +
  scale_color_viridis_c()
f <- file.path(plot_dir, paste0("pca_", method, "_bg_ks_dist.pdf"))
ggsave(f, plot=p1, device="pdf", width=width, height=height)

# patient
p1 <- pca_plot(pca_res, pca_tbl, patient, path_specimen) +
  scale_color_manual(values = color_scales$patient)
f <- file.path(plot_dir, paste0("pca_", method, "_patient.pdf"))
ggsave(f, plot=p1, device="pdf", width=width, height=height)

# path
p1 <- pca_plot(pca_res, pca_tbl, path_specimen, path_specimen) +
  scale_color_manual(values = color_scales$path_specimen)
#  scale_color_manual(values = pals::cols25())
f <- file.path(plot_dir, paste0("pca_", method, "_path_specimen.pdf"))
ggsave(f, plot=p1, device="pdf", width=width, height=height)

p1 <- pca_plot(pca_res, pca_tbl, path_aoi, path_specimen) + 
  scale_color_manual(values = color_scales$path_aoi)
# p1 <- pca_plot(pca_res, pca_tbl, path_aoi, path_specimen) +
#   scale_color_manual(values = pals::cols25())
f <- file.path(plot_dir, paste0("pca_", method, "_path_aoi.pdf"))
ggsave(f, plot=p1, device="pdf", width=width, height=height)

```


## DE analysis INV specimens

```{r de analysis inv specimens}

# DE analysis 
de <- NULL

# DE analysis of INV lesions
s <- filter(fds$samples, path_specimen == "inv")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(invinv_vs_invnl = path_aoiinv - path_aoinl,
                           invhgd_vs_invnl = path_aoihgd - path_aoinl,
                           invlgd_vs_invnl = path_aoilgd - path_aoinl,
                           invinv_vs_invlgd = path_aoiinv - path_aoilgd,
                           invhgd_vs_invlgd = path_aoihgd - path_aoilgd,
                           invinv_vs_invhgd = path_aoiinv - path_aoihgd,
                           levels=design)
y <- log2_norm_cpm[, s$aoi]
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
res <- res %>% mutate(method = method)
de <- bind_rows(de, res)

table(de$contrast, de$de)


# de genes that associate with path
inv_de_merged <- select(res, gene, contrast, avgexpr, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))

inv_de_merged <- inv_de_merged %>%
  mutate(subtype = case_when(de_invinv_vs_invnl == "up" & de_invhgd_vs_invnl == "up" & de_invinv_vs_invlgd == "up" & de_invhgd_vs_invlgd == "up" ~ "up",
                             de_invinv_vs_invnl == "dn" & de_invhgd_vs_invnl == "dn" & de_invinv_vs_invlgd == "dn" & de_invhgd_vs_invlgd == "dn" ~ "dn",                             
                             TRUE ~ "none"))

table(inv_de_merged$subtype)
nrow(filter(inv_de_merged, subtype != "none"))

# heatmap
g <- filter(inv_de_merged, subtype != "none") %>% pull(gene)
s <- fds$samples
x <- log2_norm_cpm[g, s$aoi]

# median center
rowmeds <- apply(x, 1, median)
m <- sweep(x, 1, rowmeds, FUN="-")


annot_col <- column_to_rownames(s, "aoi") %>% 
  select(patient = patient,
         path_aoi = path_aoi,
         path_specimen = path_specimen) %>%
  as.data.frame()

gene_annot <- fds$meta %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))

colnames(gene_annot)

annot_row <- filter(gene_annot, gene %in% g) %>%
  select(gene, de_cptac_rna) %>%
  column_to_rownames("gene") %>%
  as.data.frame()

table(annot_row$de_cptac_rna)
colnames(annot_row)

f <- file.path(plot_dir,  paste0("heatmap_de_invhgd.pdf"))
p <- plot_pheatmap(m, annot_col, annot_row, color_scales)
h <- round(3 + (length(g) / 10))
save_pheatmap_pdf(p, filename=f, width=10, height=h)

```

## DE analysis LGD specimens

```{r de analysis lgd specimens}

# DE analysis of LGD lesions
s <- filter(fds$samples, path_specimen == "lgd")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd"))

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(lgdlgd_vs_lgdnl = path_aoilgd - path_aoinl,
                           levels=design)
y <- log2_norm_cpm[, s$aoi]
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
res <- res %>% mutate(method = method)
table(res$contrast, res$de)
de <- bind_rows(de, res)

```

## DE analysis LGD vs INV specimens

```{r de lgd vs inv specimens}

# compare specimens with LGD vs specimens with INV
s <- fds$samples
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
s$path_specimen <- factor(s$path_specimen, levels=c("lgd", "inv"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, levels=c("lgd_nl", "lgd_lgd", "inv_nl", "inv_lgd", "inv_hgd", "inv_inv"))
levels(s$path_patient)
table(s$path_patient)

design <- model.matrix(~0 + path_patient, s)
colnames(design) <- levels(s$path_patient)
colnames(design)
contrasts <- makeContrasts(
  invinv_vs_lgdnl = inv_inv - lgd_nl,
  invhgd_vs_lgdnl = inv_hgd - lgd_nl,
  invlgd_vs_lgdnl = inv_lgd - lgd_nl,
  invnl_vs_lgdnl = inv_nl - lgd_nl,
  invinv_vs_lgdlgd = inv_inv - lgd_lgd,
  invhgd_vs_lgdlgd = inv_hgd - lgd_lgd,
  invlgd_vs_lgdlgd = inv_lgd - lgd_lgd,
  levels=design
)

y <- log2_norm_cpm[, s$aoi]

res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
res <- res %>% mutate(method = method)
de <- bind_rows(de, res)
table(res$contrast, res$de)

# de genes 
specimen_de_merged <- select(res, gene, contrast, avgexpr, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))

colnames(specimen_de_merged)
specimen_de_merged <- specimen_de_merged %>%
  mutate(subtype = case_when(de_invlgd_vs_lgdlgd == "up" & de_invnl_vs_lgdnl == "up" ~ "both_up",
                             de_invlgd_vs_lgdlgd == "dn" & de_invnl_vs_lgdnl == "dn" ~ "both_dn",
                             de_invlgd_vs_lgdlgd == "up" & de_invnl_vs_lgdnl == "no" ~ "lgd_up",
                             de_invlgd_vs_lgdlgd == "dn" & de_invnl_vs_lgdnl == "no" ~ "lgd_dn",
                             de_invlgd_vs_lgdlgd == "no" & de_invnl_vs_lgdnl == "up" ~ "nl_up",
                             de_invlgd_vs_lgdlgd == "no" & de_invnl_vs_lgdnl == "dn" ~ "nl_dn",  
                             TRUE ~ "none"))

table(specimen_de_merged$subtype)
nrow(filter(specimen_de_merged, subtype != "none"))

# heatmap
g <- filter(specimen_de_merged, subtype != "none") %>% pull(gene)
s <- fds$samples %>% filter(path_aoi %in% c("lgd", "nl"))
x <- log2_norm_cpm[g, s$aoi]

annot_col <- column_to_rownames(s, "aoi") %>% 
  select(patient = patient,
         path_aoi = path_aoi,
         path_specimen = path_specimen) %>%
  as.data.frame()

gene_annot <- fds$meta %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))

annot_row <- filter(gene_annot, gene %in% g) %>%
  select(gene, de_cptac_rna) %>%
  column_to_rownames("gene") %>%
  as.data.frame()

table(annot_row$de_cptac_rna)
colnames(annot_row)

f <- file.path(plot_dir,  paste0("heatmap_de_specimen.pdf"))
p <- plot_pheatmap(x, annot_col, annot_row, color_scales)
h <- round(3 + (length(g) / 10))
save_pheatmap_pdf(p, filename=f, width=10, height=h)


#
# 2D scatter plot
#
x <- specimen_de_merged
nrow(specimen_de_merged %>% filter(subtype != "none"))
table(specimen_de_merged$subtype)

color_scale = c(none = "#888888",
                both_up = "#FF0000",
                both_dn = "#0000FF",
                nl_up = "#32cd32",
                nl_dn = "#670092",
                lgd_dn = "#1677c4",
                lgd_up = "#ff6700")

size_scale = c(none = 1,
               both_up = 2,
               both_dn = 2,
               nl_up = 2,
               nl_dn = 2,
               lgd_dn = 2,
               lgd_up = 2)

annot_data <- x %>%
  filter(subtype != "none")
none_data <- x %>%
  filter(subtype == "none")

label_data <- filter(x, subtype != "none") %>%
  mutate(log2fc_mean = mean(c_across(c("log2fc_invlgd_vs_lgdlgd", "log2fc_invnl_vs_lgdnl"))))

label_data <- bind_rows(filter(label_data, subtype == "both_up") %>% slice_max(log2fc_mean, n = 10),
                        filter(label_data, subtype == "both_dn") %>% slice_min(log2fc_mean, n = 10),
                        filter(label_data, subtype == "nl_up") %>% slice_max(log2fc_invnl_vs_lgdnl, n = 10),
                        filter(label_data, subtype == "nl_dn") %>% slice_min(log2fc_invnl_vs_lgdnl, n = 10),
                        filter(label_data, subtype == "lgd_dn") %>% slice_min(log2fc_invlgd_vs_lgdlgd, n = 10),
                        filter(label_data, subtype == "lgd_up") %>% slice_max(log2fc_invlgd_vs_lgdlgd, n = 10))
label_data <- distinct(label_data)
nrow(label_data)

p <- ggplot(x, aes(x=log2fc_invnl_vs_lgdnl, 
                   y=log2fc_invlgd_vs_lgdlgd,
                   color=subtype, 
                   size=subtype)) +
  geom_point(data=none_data, alpha=0.4) + 
  geom_point(data=annot_data, alpha=0.8) +
  geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +
  scale_color_manual(values = color_scale) +
  scale_size_manual(values = size_scale, guide = "none") +
  theme_minimal() +
  ylab("log2(Fold Change) INV-LGD vs LGD-LGD") +
  xlab("log2(Fold Change) INV-NL vs LGD-NL") +
  labs(color = "DE Genes")
p
f <- file.path(plot_dir, "scatter_de_specimen.pdf")
ggsave(f, plot=p, device="pdf", width=8, height=8)




table(de$contrast)
write_xlsx(de, file.path(working_dir, "de_path.xlsx"))


```

## 2D scatter plots

```{r 2d scatter plots}

# LGD vs NL
de_merged <- select(de, gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))
colnames(de_merged)
de_merged <- de_merged %>%
  mutate(subtype = case_when(de_invlgd_vs_invnl == "up" & de_lgdlgd_vs_lgdnl == "up" ~ "both_up",
                             de_invlgd_vs_invnl == "dn" & de_lgdlgd_vs_lgdnl == "dn" ~ "both_dn",
                             de_invlgd_vs_invnl == "up" & de_lgdlgd_vs_lgdnl == "dn" ~ "inv_up_lgd_dn",
                             de_invlgd_vs_invnl == "dn" & de_lgdlgd_vs_lgdnl == "up" ~ "inv_dn_lgd_up", 
                             de_invlgd_vs_invnl == "up" & de_lgdlgd_vs_lgdnl == "no" ~ "inv_up",
                             de_invlgd_vs_invnl == "dn" & de_lgdlgd_vs_lgdnl == "no" ~ "inv_dn",
                             de_invlgd_vs_invnl == "no" & de_lgdlgd_vs_lgdnl == "up" ~ "lgd_up",
                             de_invlgd_vs_invnl == "no" & de_lgdlgd_vs_lgdnl == "dn" ~ "lgd_dn",  
                             TRUE ~ "none"))
table(de_merged$subtype)
nrow(filter(de_merged, subtype != "none"))

de_merged %>% filter(subtype == "inv_dn_lgd_up") %>% as.data.frame()

#
# 2D scatter plot
#
x <- de_merged
nrow(de_merged %>% filter(subtype != "none"))
table(de_merged$subtype)

color_scale = c(none = "#888888",
                both_up = "#FF0000",
                both_dn = "#0000FF",
                inv_up_lgd_dn = "#FFFF00",
                inv_dn_lgd_up = "#00ffff",
                inv_up = "#32cd32",
                inv_dn = "#670092",
                lgd_dn = "#1677c4",
                lgd_up = "#ff6700")

size_scale = c(none = 1,
               both_up = 2,
               both_dn = 2,
               inv_up_lgd_dn = 2,
               inv_dn_ldg_up = 2,
               inv_up = 2,
               inv_dn = 2,
               lgd_dn = 2,
               lgd_up = 2)

annot_data <- x %>%
  filter(subtype != "none")
none_data <- x %>%
  filter(subtype == "none")

label_data <- filter(x, subtype != "none") %>%
  rowwise() %>%
  mutate(log2fc_mean = mean(c_across(c("log2fc_invlgd_vs_invnl", "log2fc_lgdlgd_vs_lgdnl")))) %>%
  ungroup()

label_data <- bind_rows(filter(label_data, subtype == "both_up") %>% slice_max(log2fc_mean, n = 10),
                        filter(label_data, subtype == "both_dn") %>% slice_min(log2fc_mean, n = 10),
                        filter(label_data, subtype == "inv_up_lgd_dn") %>% slice_max(log2fc_mean, n = 10),
                        filter(label_data, subtype == "inv_dn_lgd_up") %>% slice_min(log2fc_mean, n = 10),
                        filter(label_data, subtype == "inv_up") %>% slice_max(log2fc_invlgd_vs_invnl, n = 10),
                        filter(label_data, subtype == "inv_dn") %>% slice_min(log2fc_invlgd_vs_invnl, n = 10),
                        filter(label_data, subtype == "lgd_up") %>% slice_max(log2fc_lgdlgd_vs_lgdnl, n = 10),
                        filter(label_data, subtype == "lgd_dn") %>% slice_min(log2fc_lgdlgd_vs_lgdnl, n = 10))
label_data <- distinct(label_data)
nrow(label_data)
table(label_data$subtype)

p <- ggplot(x, aes(x=log2fc_lgdlgd_vs_lgdnl, 
                   y=log2fc_invlgd_vs_invnl,
                   color=subtype, 
                   size=subtype)) +
  geom_point(data=none_data, alpha=0.4) + 
  geom_point(data=annot_data, alpha=0.8) +
  geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +
  scale_color_manual(values = color_scale) +
  scale_size_manual(values = size_scale, guide = "none") +
  theme_minimal() +
  xlab("log2(Fold Change) LGD-LGD vs LGD-NL") +
  ylab("log2(Fold Change) INV-LGD vs INV-NL") +
  labs(color = "DE Genes")
p

f <- file.path(plot_dir, "scatter_de_inv_vs_lgd.pdf")
ggsave(f, plot=p, device="pdf", width=8, height=8)


```


## DE Plots

```{r de plots}

table(de$contrast)

x <- filter(de, contrast == "invlgd_vs_invnl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "volcano_de_invlgd_vs_invnl.pdf"), p, width=8, height=8)
ggsave(file.path(plot_dir, "volcano_de_invlgd_vs_invnl.png"), p, width=8, height=8)

x <- filter(de, contrast == "invlgd_vs_lgdlgd")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "volcano_de_invlgd_vs_lgdlgd.pdf"), p, width=8, height=8)
ggsave(file.path(plot_dir, "volcano_de_invlgd_vs_lgdlgd.png"), p, width=8, height=8)

x <- filter(de, contrast == "invnl_vs_lgdnl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "volcano_de_invnl_vs_lgdnl.pdf"), p, width=8, height=8)
ggsave(file.path(plot_dir, "volcano_de_invnl_vs_lgdnl.png"), p, width=8, height=8)

x <- filter(de, contrast == "lgdlgd_vs_lgdnl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "volcano_de_lgdlgd_vs_lgdnl.pdf"), p, width=8, height=8)
ggsave(file.path(plot_dir, "volcano_de_lgdlgd_vs_lgdnl.png"), p, width=8, height=8)



x <- filter(de, contrast == "invhgd_vs_lgdlgd")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "de_invhgd_vs_lgdlgd.pdf"), p, width=8, height=8)
ggsave(file.path(plot_dir, "de_invhgd_vs_lgdlgd.png"), p, width=8, height=8)

x <- filter(de, contrast == "invhgd_vs_lgdnl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "de_invhgd_vs_lgdnl.pdf"), p, width=8, height=8)
ggsave(file.path(plot_dir, "de_invhgd_vs_lgdnl.png"), p, width=8, height=8)








# volcano plots
x <- filter(de, contrast == "inv_vs_nl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "de_inv_vs_nl.pdf"), p, width=8, height=8)

x <- filter(de, contrast == "hgd_vs_nl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "de_hgd_vs_nl.pdf"), p, width=8, height=8)

x <- filter(de, contrast == "lgd_vs_nl")
table(x$de)
p <- plot_de_volcano(x)
ggsave(file.path(plot_dir, "de_lgd_vs_nl.pdf"), p, width=8, height=8)





```

## GSEA

```{r gsea}

get_de_ranks <- function(de, a) {
  ranks <- de %>% 
    filter(contrast == a) %>%
    select(gene, log2fc, padj) %>%
    mutate(rank = log2fc)
  ranks = sort(setNames(ranks$rank, ranks$gene), decreasing = TRUE)
  return(ranks)
}

plot_gsea_enrichment <- function(x, ranks, a, gs) {
  txt_stat <- paste0("ES=", round(x$ES,2), " NES=", round(x$NES, 2), " padj=", format(x$padj, scientific=TRUE, digits=3))
  txt_title <- paste0("ranks: ", a, " gs: ", x$pathway)
  p <- plotEnrichment(gs[[x$pathway]], ranks) +
    annotate(geom="text", x=150, y=0.1, label=txt_stat, hjust=0) +
    labs(title = txt_title, 
         xlab = "Rank",
         ylab = "Enrichment Score") +
    theme(plot.title = element_text(size=8))
  return(p)
}

run_batch_fgsea <- function(my_analyses, my_de, my_gs, my_prefix, my_plot_dir, my_padj_cutoff=0.01) {
  gsea <- NULL
  for (a in my_analyses) {
    ranks <- get_de_ranks(my_de, a)
    res <- fgsea(pathways = my_gs, stats = ranks, minSize = 10, eps = 0, nPermSimple = 10000)
    res <- mutate(res, analysis = a)
    gsea <- bind_rows(gsea, res)
    print(a)
    for (i in 1:nrow(res)) {
      x <- res[i,]
      if (x$padj >= my_padj_cutoff) { next }
      print(x$pathway)
      p <- plot_gsea_enrichment(x, ranks, a, my_gs)
      f <- file.path(my_plot_dir, paste0(my_prefix, "_", a, "_gs_", x$pathway, ".pdf"))
      ggsave(f, plot=p, device="pdf", width=5, height=3)
    }
  }
  return(gsea)
}

gmt_to_tbl <- function(gmt) {
  gs_to_tbl <- function(gs_name, gs_genes) {
    bind_cols(gs_name = rep(gs_name, length(gs_genes)),
              gene_symbol = gs_genes)
  }
  tbl <- NULL
  for (i in 1:length(gmt)) {
    tbl <- bind_rows(tbl, gs_to_tbl(names(gmt)[i], gmt[[i]]))
  }
  return(tbl)
}

plot_gsea_volcano <- function(x, padj_cutoff = 0.01, max.overlaps = Inf) {
  p <- ggplot(x, aes(x=NES, y=-log10(padj), color=analysis)) + 
    geom_point(data=subset(x, padj > padj_cutoff), size=1, alpha=0.4) + 
    geom_point(data=subset(x, padj <= padj_cutoff), size=2, alpha=0.8) +
    geom_text_repel(data=subset(x, padj <= padj_cutoff), color="black", size=3, aes(label=pathway), max.overlaps=max.overlaps) +
    ylab("-log10(adjusted p-value)") +
    xlab("NES") +
    theme_minimal() +
    theme(legend.position="bottom") + 
    theme(axis.line = element_line(color = "black")) +
    coord_flip() 
  return(p)
}

plot_gsea_barplot <- function(x, padj_cutoff = 0.01) {
  x <- mutate(x, sig = ifelse(padj < padj_cutoff, ifelse(NES < 0, "dn", "up"), "no"),
              neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))
  p <- ggplot(x, aes(x=reorder(pathway, NES), y=NES, fill=neglog10padj)) +
    geom_col() +
    scale_fill_gradient(low = "blue", high = "cyan", na.value="#aaaaaa") +
    coord_flip() +
    labs(x="Analysis", y="Normalized Enrichment Score") +
    theme_minimal()
  return(p)
}

gsea_leading_edge_matrix <- function(x, my_padj_cutoff) {
  # filter for significant results
  sig_pathways <- x %>%
    filter(padj < my_padj_cutoff) %>% select(pathway) %>% distinct() %>% pull()
  x <- x %>%
    filter(pathway %in% sig_pathways) %>%
    arrange(desc(NES))
  
  # get leading edge genes
  leading_edge_genes <- summarise(x, result = unlist(leadingEdge, use.names=FALSE)) %>% distinct() %>% pull()
  
  # matrix genes (rows) vs pathways (columns)
  m <- matrix(data = 0, nrow=length(leading_edge_genes), ncol=nrow(x))
  m <- as.data.frame(m)
  rownames(m) <- leading_edge_genes
  colnames(m) <- x$pathway
  for (i in 1:nrow(x)) {
    p <- x$pathway[i]
    nes <- x$NES[i]
    padj <- x$padj[i]
    for (j in x$leadingEdge[i]) {
      m[j, p] <- sign(nes)
    }
  }
  # filter genes only found in a single pathway
  m <- m[abs(rowSums(m)) >= 2, ]
  # filter empty columns
  m <- m[, abs(colSums(m)) > 1]
  # order by number of pathways sharing the gene
  m <- m[order(rowSums(m), decreasing=TRUE),]
  m <- m[,order(colSums(m), decreasing=TRUE)]
  return(m)
}


#
# Setup analysis
#
table(de$contrast)
analyses <- unique(de$contrast)

# CPTAC GSEA
prefix <- "gsea_cptac"
padj_cutoff <- 0.01
gs <- gs_pdac[c("CPTAC_RNA_ALL_UP", "CPTAC_RNA_ALL_DN")]
gsea <- run_batch_fgsea(analyses, de, gs, prefix, gsea_plot_dir, padj_cutoff)
gsea_cptac <- as_tibble(gsea)

# write gsea results
f <- file.path(working_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(working_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)


#
# Hallmark gene set analysis
#
prefix <- "gsea_hallmark"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, de, gs_hallmark, prefix, gsea_plot_dir, padj_cutoff)
gsea_hallmark <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_subcat, gs_name) %>% distinct(), 
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(working_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(working_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# Gene Ontology analysis
#
# prefix <- "gsea_gobp"
# padj_cutoff <- 0.01
# gsea <- run_batch_fgsea(analyses, de, gs_gobp, prefix, gsea_plot_dir, padj_cutoff)
# gsea_gobp <- as_tibble(gsea) %>%
#   left_join(select(msigdb_gene_sets, gs_cat, gs_subcat, gs_name) %>% distinct(), 
#             by=c("pathway"="gs_name"))
# # write gsea results
# f <- file.path(working_dir, paste0(prefix, "_results.tsv"))
# data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
# f <- file.path(working_dir, paste0(prefix, "_results.xlsx"))
# write_xlsx(gsea, f)

#
# GSEA hallmark bar plots
#

table(gsea_hallmark$analysis)
x <- gsea_hallmark
x$pathway <- gsub("HALLMARK_", "", x$pathway)

p <- plot_gsea_barplot(filter(x, analysis == "invinv_vs_invnl"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invinv_vs_invnl_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invinv_vs_invlgd"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invinv_vs_invlgd_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invhgd_vs_invnl"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invhgd_vs_invnl_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invhgd_vs_invlgd"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invhgd_vs_invlgd_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invlgd_vs_invnl"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invlgd_vs_invnl_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invlgd_vs_lgdnl"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invlgd_vs_lgdnl_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invlgd_vs_lgdlgd"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invlgd_vs_lgdlgd_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "invnl_vs_lgdnl"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_invnl_vs_lgdnl_hallmark_barplot.pdf"), plot=p, width=6, height=6)

p <- plot_gsea_barplot(filter(x, analysis == "lgdlgd_vs_lgdnl"), padj_cutoff = 0.01)
ggsave(file.path(plot_dir, "gsea_lgdlgd_vs_lgdnl_hallmark_barplot.pdf"), plot=p, width=6, height=6)





#
# cptac plots comparing analysis
#
unique(gsea_cptac$analysis)
padj_cutoff <- 0.01
plotanalyses <- unique(gsea_cptac$analysis)

inv_analyses <- c("inv_vs_nl", "hgd_vs_nl", "lgd_vs_nl", "inv_vs_lgd", "hgd_vs_lgd", "inv_vs_hgd")
lgd_analyses <- c("lgdlgd_vs_lgdnl")
control_analyses <- c("invinv_vs_lgdnl", "invhgd_vs_lgdnl", "invlgd_vs_lgdnl",
                      "invinv_vs_lgdlgd", "invhgd_vs_lgdlgd")
mix_analyses <- c("invlgd_vs_lgdlgd", "invnl_vs_lgdnl")

x <- filter(gsea_cptac, analysis %in% plotanalyses)
x$analysis <- factor(x$analysis, levels=plotanalyses)
x <- x %>% mutate(
  facet = case_when(analysis %in% inv_analyses ~ "inv",
                    analysis %in% lgd_analyses ~ "lgd",
                    analysis %in% mix_analyses ~ "invlgd_same",
                    analysis %in% control_analyses ~ "control")
)

x <- x %>% mutate( 
  sig = ifelse(padj < padj_cutoff, ifelse(NES < 0, "dn", "up"), "no"),
  neglog10padj = -log10(x$padj),
  score = NES * -log10(x$padj),
  fillneglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  fillnes = ifelse(padj < padj_cutoff, NES, NA),
)

p <- ggplot(x, aes(x=reorder(analysis, abs(NES)), y=NES, fill=fillneglog10padj)) +
  geom_bar(stat="identity", position=position_dodge()) +
#  coord_flip() +
  scale_fill_viridis_c() +
#  scale_fill_gradient(low = "#004488", high = "cyan", na.value="#aaaaaa") +
  labs(x="Analysis", y="Normalized Enrichment Score",
       title="PDAC DE Genes (CPTAC RNA-Seq)") + 
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1, size=6)) +
  facet_wrap(~ facet + pathway, nrow=1, scales="free_x")
p
ggsave(file.path(plot_dir, "gsea_cptac_barplots.pdf"), p, width=9, height=5)



```


## Plot individual genes

```{r plot individual genes}

plot_gene <- function(s, x, gene, aes_x, aes_color) {
  exprs <- unlist(x[gene, ])
  x <- s %>% add_column(gene = exprs)
  p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
    geom_boxplot(outlier.shape = NA, width=0.75) +
#    geom_violin(scale="width", width=0.5) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
    ggtitle(paste0("Gene: ", gene)) +
    theme_minimal() + 
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
    theme(axis.line = element_line(color = "black"))
  return(p)
}


y <- log2_norm_cpm
s <- fds$samples
s <- fds$samples
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
s$path_specimen <- factor(s$path_specimen, levels=c("lgd", "inv"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, levels=c("lgd_nl", "lgd_lgd", "inv_nl", "inv_lgd", "inv_hgd", "inv_inv"))
levels(s$path_patient)
table(s$path_patient)

x <- specimen_de_merged
arrange(specimen_de_merged, log2fc_invlgd_vs_lgdnl) %>% as.data.frame()
x$sumlog2fc <- specimen_de_merged$log2fc_invlgd_vs_lgdnl
width <- 5
height <- 6

p <- plot_gene(s, y, "MUC5AC", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_MUC5AC.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "S100P", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_S100P.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "CEACAM5", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_CEACAM5.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "SYTL2", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_SYTL2.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "GALNT6", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_GALNT6.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "CELA3A", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_CELA3A.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "PNLIP", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_PNLIP.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "AMY1A", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_AMY1A.pdf"), p, width=width, height=height)
  
p <- plot_gene(s, y, "BPIFB1", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_BPIFB1.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "EPPK1", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_EPPK1.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "PITX1", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_PITX1.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "CTRB1", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
ggsave(file.path(plot_dir, "boxplot_gene_CTRB1.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "CLU", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
p
ggsave(file.path(plot_dir, "boxplot_gene_CLU.pdf"), p, width=width, height=height)

p <- plot_gene(s, y, "NKX6-2", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
p
ggsave(file.path(plot_dir, "boxplot_gene_NKX6-2.pdf"), p, width=width, height=height)


p <- plot_gene(s, y, "OLFM4", path_patient, path_patient) + 
  scale_fill_manual(values=color_scales$path_patient)
p

specimen_de_merged %>% filter(gene == "CD55")

arrange(log2fc_invlgd_vs_lgdnl)

inv_de_merged %>% arrange()
specimen_de_merged %>% arrange(log2fc_invlgd_vs_lgdnl)

colnames(inv_de_merged)



```



## Compare background correction approaches

```{r bgcorrect compare}

normcounts = list()
bgcorrect_methods <- c("none", "qq", "bgsub", "kdeppv", "kdefor")
norm_methods <- c("cpm", "qn", "rle", "q3", "tmm")
pca_bg_fits <- NULL
de <- NULL

# setup for de analysis
s <- filter(fds$samples, path_specimen == "inv")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(inv_vs_nl = path_aoiinv - path_aoinl,
                           hgd_vs_nl = path_aoihgd - path_aoinl,
                           lgd_vs_nl = path_aoilgd - path_aoinl,
                           levels=design)


for (bgcorrect_method in bgcorrect_methods) {
  for (norm_method in norm_methods) {
    method <- paste0(bgcorrect_method, "_", norm_method)
    print(method)
    x <- st_geomx_bgcorrect(fds, method=bgcorrect_method, 
                            bw.adjust=bgcorrect_bw_adjust, 
                            bg.quant=bgcorrect_bg_quant)
    x <- st_geomx_normalize(fds, x, method=norm_method)
    
    # only most variable genes
    y <- most_cv_genes(log2(x), 5000)
    rsq <- pca_bg_fit(y, fds$samples$bg_ks_dist, npcs=50)
    pca_bg_fits <- bind_rows(pca_bg_fits,
                             bind_cols(bgcorrect_method=bgcorrect_method, 
                                       norm_method=norm_method,
                                       rsq=rsq))

    ## PCA analysis
    pca_res <- run_pca(y)
    pca_tbl <- bind_cols(fds$samples, as_tibble(pca_res$pca$x))

    # bg_ks_dist
    p1 <- pca_plot(pca_res, pca_tbl, bg_ks_dist, path_specimen) +
      scale_color_viridis_c()
    f <- file.path(plot_dir, paste0("pca_", method, "_bg_ks_dist.pdf"))
    ggsave(f, plot=p1, device="pdf")

    # patient
    p1 <- pca_plot(pca_res, pca_tbl, patient, path_specimen) +
      scale_color_manual(values = pals::cols25())
    f <- file.path(plot_dir, paste0("pca_", method, "_patient.pdf"))
    ggsave(f, plot=p1, device="pdf")

    # path
    p1 <- pca_plot(pca_res, pca_tbl, path_specimen, path_specimen) +
      scale_color_manual(values = pals::cols25())
    f <- file.path(plot_dir, paste0("pca_", method, "_path_specimen.pdf"))
    ggsave(f, plot=p1, device="pdf")

    p1 <- pca_plot(pca_res, pca_tbl, path_aoi, path_specimen) +
      scale_color_manual(values = pals::cols25())
    f <- file.path(plot_dir, paste0("pca_", method, "_path_aoi.pdf"))
    ggsave(f, plot=p1, device="pdf")

    ## DE analysis
    y <- log2(x[,s$aoi])
    res <- run_limma_trend(y, design, contrasts, 
                           padj_cutoff=de_padj_cutoff, 
                           log2fc_cutoff=de_log2fc_cutoff)
    res <- res %>% mutate(
      bgcorrect_method = bgcorrect_method,
      norm_method = norm_method
    )
    de <- bind_rows(de, res)
  }
}

de <- de %>% mutate(
  analysis = paste0(contrast, "_", bgcorrect_method, "_", norm_method)
)


# gsea
de_inv <- filter(de, contrast == "inv_vs_nl")
gsea <- NULL
fgsea_nperm <- 10000


ranks <- get_de_ranks(de_inv, "inv_vs_nl_none_cpm", "lfc")
res <- fgseaSimple(pathways=gs_pdac,
             stats=ranks,
             nperm=100000,
             nproc=1)

range(ranks)
summary(ranks)
sum(is.na(ranks))
length(unique(names(ranks)))
#                 nproc=1)


    ranks <- get_de_ranks(de, a, rank_method)
#     res <- fgseaSimple(pathways=gs,
#                        stats=ranks,
#                        nperm=fgsea.nperm)
# #                       nproc=1)
    res <- fgsea(pathways=gs,
                 stats=ranks,
                 eps=0,
                 nPermSimple=fgsea.nperm)
#                 nproc=1)

for (rank_method in c("lfc", "p", "lfcp", "q")) {
  print(rank_method)
  res <- run_batch_fgsea(gs_pdac, de_inv, unique(de_inv$analysis), 
                         rank_method=rank_method, 
                         fgsea.nperm=fgsea_nperm)
  res$rank_method <- rank_method
  gsea <- bind_rows(gsea, res)
}

gsea <- gsea %>% 
  rowwise() %>% 
  mutate(num_leading_edge = length(leadingEdge)) %>% 
  ungroup()

gsea_pdac <- as_tibble(gsea)

# write comparison data
write_xlsx(list(de = de, 
                gsea_pdac = gsea_pdac,
                pca_bg_fits = pca_bg_fits),
           file.path(working_dir, "bg_correct_cmp_results.xlsx"))
# write gsea results
f <- file.path(working_dir, paste0("gsea_pdac_results.tsv"))
data.table::fwrite(gsea_pdac, file=f, sep="\t", sep2=c("", " ", ""))




p <- plot_gsea_barplot(gsea) +
  facet_grid(rank_method ~ analysis)

ggsave(file.path(plot_dir, "gsea_bg_cmp_pdac.pdf"), p, width=30, height=20)

?ggsave
plot_gsea_volcano(gsea_pdac)

# PDAC GSEA Analysis
prefix <- "gsea_pdac"
padj_cutoff <- 0.05
run_batch_fgsea(analyses)
gsea <- run_batch_fgsea(analyses, de, gs_pdac)






```

## QC: Background correction

```{r background correction}

# params for plot
quantiles=c(0.10, 0.25, 0.5, 0.75, 0.90, 0.99)
width <- 16
height <- 6

# params for bgcorrect
bw.adjust <- 1
bg.quant <- 0.5


## plots

# raw cpm
x <- as_tibble(cpm(ds$counts))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_none_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_none_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, xraw)
# f <- file.path(plot_dir, "qc_bgcorrect_none_boxplot.pdf")
# ggsave(f, p, width=width, height=8)

# bgsub cpm
x <- st_geomx_bgcorrect(ds, method="bgsub", bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_bgsub_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_bgsub_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_bgsub_boxplot.pdf")
# ggsave(f, p, width=16, height=8)

# qq cpm
x <- st_geomx_bgcorrect(ds, method="qq", bw.adjust=bw.adjust, bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_qq_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_qq_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_qq_boxplot.pdf")
# ggsave(f, p, width=16, height=8)

# ppv cpm
x <- st_geomx_bgcorrect(ds, method="kdeppv", bw.adjust=bw.adjust, bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_ppv_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_ppv_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_ppv_boxplot.pdf")
# ggsave(f, p, width=16, height=8)

# for cpm
x <- st_geomx_bgcorrect(ds, method="kdefor", bw.adjust=bw.adjust, bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_kdefor_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_kdefor_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_kdefor_boxplot.pdf")
# ggsave(f, p, width=16, height=8)


```



## Apply filters and normalize

Consolidate technical replicates. When sample has replicates, choose replicate with greater signal-to-noise ratio

```{r filter and normalize}


# bg correct
#x <- fds$counts

x <- st_geomx_bgcorrect(fds, method=bgcorrect_method, 
                        bw.adjust=bgcorrect_bw_adjust, 
                        bg.quant=bgcorrect_bg_quant)

# normalize
normcounts <- st_geomx_normalize(fds, x, method=norm_method)

```



## QC plot: normalization

```{r qc normalization}

# sample annotations to use in plots
s <- select(fds$samples, aoi, slide, keep, path_aoi, path_specimen)
x <- bind_cols(fds$meta, normcounts)
x <- x %>%
  pivot_longer(s$aoi, names_to="aoi", values_to="count") %>%
  inner_join(s, by=c("aoi"="aoi"), suffix=c("_gene", "_aoi"))

y <- x %>%
  group_by(aoi, expressed) %>%
  summarise(slide = unique(slide),
            med = median(count),
            q3 = quantile(count, 0.75)) %>%
  ungroup()

p <- ggplot(x, aes(x=count, y=reorder(aoi, count), fill=slide)) +
  stat_density_ridges(color="#00000066", alpha=0.4, scale=10, rel_min_height=0.001, quantile_lines=TRUE) +
  scale_x_log10() +
  scale_fill_manual(values = pals::cols25()) +
  xlab("Count") +
  ylab("AOIs") +
  labs(fill = "Slide") +
  theme_ridges() +
  theme(axis.text.y = element_blank()) +
  facet_grid(slide ~ ., scales="free_y")
p
f <- file.path(plot_dir, "qc_ridge_plot_norm_aoi.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)


# by aoi path
p <- x %>%
  ggplot(aes(x=count, y=factor(path_aoi), fill=factor(path_aoi))) +
  stat_density_ridges(alpha=0.7, scale=3, rel_min_height=0.001, quantile_lines=TRUE) + 
  scale_color_manual(values = pals::cols25()) + 
  scale_x_log10() +
  theme_ridges() + 
  theme(axis.text = element_text(size=5))
p
f <- file.path(plot_dir, "qc_ridge_plot_norm_path.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)

# by specimen path
p <- x %>%
  ggplot(aes(x=count, y=factor(path_aoi), fill=factor(path_specimen))) +
  stat_density_ridges(alpha=0.7, scale=3, rel_min_height=0.001, quantile_lines=TRUE) + 
  scale_color_manual(values = pals::cols25()) + 
  scale_x_log10() +
  theme_ridges() + 
  theme(axis.text = element_text(size=5))
p
f <- file.path(plot_dir, "qc_ridge_plot_norm_class.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)


```


## Unsupervised clustering / dimensionality reduction

```{r dim reduction}
library(factoextra)
library(umap)

most_cv_genes <- function(x, ntop=500) {
  rv <- apply(x, 1, function(x) sd(x) / mean(x))
  keep <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  return(x[keep,])
}

run_pca <- function(x) {
  pca <- prcomp(t(x))
  pct_var <- round(100 * pca$sdev^2 / sum( pca$sdev^2 ), 2)
  pct_var <- paste(colnames(pca$x), " (", paste(as.character(pct_var), "%", ")", sep=""), sep="")
  return(list(pca = pca, pct_var = pct_var))
}

y <- log2(normcounts)
y <- most_cv_genes(y, 1000)

pca_res <- run_pca(y)
pca_tbl <- bind_cols(fds$samples, as_tibble(pca_res$pca$x))

# pca plots
p <- ggplot(pca_tbl, aes(x=PC1, y=PC2)) +
  theme_minimal() +
  labs(x=pca_res$pct_var[1], y=pca_res$pct_var[2])

# qc factors
p1 <- p + geom_point(aes(color=bg_auc), alpha = 0.8, size=2) +
  scale_color_viridis_c()
p1
ggsave(file.path(plot_dir, "pca_bg_auc.pdf"), plot=p1, device="pdf")

# segment
p1 <- p + geom_point(aes(color=segment), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1
ggsave(file.path(plot_dir, "pca_segment.pdf"), plot=p1, device="pdf")

# slide
p1 <- p + geom_point(aes(color=slide), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1
ggsave(file.path(plot_dir, "pca_path_slide.pdf"), plot=p1, device="pdf")

# path
p1 <- p + geom_point(aes(color=path_aoi, shape=path_specimen), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1
ggsave(file.path(plot_dir, "pca_path_segment.pdf"), plot=p1, device="pdf")


# investigate contributions to pca
res.pca <- prcomp(y, scale = TRUE)
p <- fviz_pca_ind(res.pca, select.ind=list(contrib=20), repel=TRUE, ggrepel.max.overlaps=Inf)
p
ggsave(file.path(plot_dir, "pca_ind.pdf"), p)

p <- fviz_contrib(res.pca, choice = "ind", axes = 1, top=20, max.overlaps=Inf)
p
ggsave(file.path(plot_dir, "pca_contrib.pdf"), p)


# UMAP variable genes
umap_fit <- t(y) %>%
   scale() %>% 
   umap()
umap_df <- umap_fit$layout %>%
   as.data.frame() %>%
   dplyr::rename(umap1 = "V1",
                 umap2 = "V2")
umap_df <- bind_cols(fds$samples, umap_df)


p <- ggplot(umap_df, aes(x=umap1, y=umap2)) +
  theme_minimal()

p1 <- p + geom_point(aes(color=bg_auc), alpha = 0.8, size=2) +
  scale_color_viridis_c()
p1

# segment
p1 <- p + geom_point(aes(color=segment), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1

# slide
p1 <- p + geom_point(aes(color=slide), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1

# segment / path
p1 <- p + geom_point(aes(color=path_aoi, shape=path_specimen), alpha = 0.8, size=2) +
  scale_color_manual(values = pals::cols25())
p1



```

## Clustering high variance genes

```{r}
library(pheatmap)

y <- log2(normcounts)
y <- most_cv_genes(y, 1000)

s <- fds$samples %>% mutate(noise_quantile = ntile(desc(bg_auc), 4))

annot_col <- column_to_rownames(s, "aoi") %>% 
  select(slide = slide,
         path_aoi = path_aoi,
         path_specimen = path_specimen,
         noise_quantile = noise_quantile) %>%
  as.data.frame()

p <- pheatmap(y,
         scale="row",
         show_rownames=FALSE,
         show_colnames=FALSE,
         border_color=NA,
         clustering_method = "average",
         clustering_distance_rows = "correlation",
         clustering_distance_cols = "correlation",
         breaks = seq(-3, 3, 0.05),
         color = colorRampPalette(c("purple3", "black", "yellow2"))(120),
         annotation_col = annot_col)

f <- file.path(plot_dir, "heatmap_cluster_highcv_genes.pdf")
save_pheatmap_pdf(p, filename=f, width=12, height=8)

```

## Differential expression analysis

```{r de analysis setup}

limma_rename_result_cols <- function(res) {
  as_tibble(res, rownames="gene") %>%
    select(gene,
           log2fc=logFC,
           avgexpr=AveExpr,
           pval=P.Value,
           padj=adj.P.Val)
}

de_format_results <- function(res, analysis, method, 
                              padj_cutoff, log2fc_cutoff) {
  mutate(res, 
         de = case_when(padj > padj_cutoff ~ "no",
                        log2fc < -log2fc_cutoff ~ "dn",
                        log2fc > log2fc_cutoff ~ "up",
                        TRUE ~ "no"),
         analysis = analysis,
         method = method)
}

run_limma_trend <- function(y, design, contrasts) {
  fit <- lmFit(y, design)
  fit <- contrasts.fit(fit, contrasts)
  fit <- eBayes(fit, trend=TRUE)
  return(fit)
}

process_limma <- function(fit, contrasts, method, 
                          padj_cutoff, log2fc_cutoff) {
  x <- NULL
  for (coef in colnames(contrasts)) {
    res <- topTable(fit, coef=coef, number=Inf, sort.by="none")
    res <- limma_rename_result_cols(res)
    res <- de_format_results(res, 
                             analysis=coef, 
                             method=method,
                             padj_cutoff=padj_cutoff, 
                             log2fc_cutoff=log2fc_cutoff)
    x <- bind_rows(x, res)
  }
  return(x)
}


plot_de_volcano <- function(res) {
  # res <- res %>% 
  #   mutate(de = case_when(padj > 0.05 ~ "no",
  #                         log2fc < 0 ~ "dn",
  #                         log2fc > 0 ~ "up"))
  p <- ggplot(res, aes(x=log2fc, y=-log10(padj), color=de, size=de)) +
    geom_point(alpha=0.7) +
#    geom_text_repel(data=subset(res, de != "no"), color="black", size=3, aes(label=gene), max.overlaps=Inf) +
    geom_text_repel(data=subset(res, de != "no"), color="black", size=3, aes(label=gene)) +
    scale_color_manual(values=c("no"="grey", "dn"="blue", "up"="red")) +
    scale_size_manual(values=c("no"=1, "dn"=2, "up"=2)) +
    theme_minimal() +
    theme(legend.position="bottom") + 
    theme(axis.line = element_line(color = "black"))
    labs(x="log2fc", y="-log10(padj)")
  return(p)
}


```


## DE analysis: epithelial compartment, specimens with invasive carcinoma

```{r between de analysis pathology}


# de <- NULL
# for (analysis in names(analysis_samples)) {
#   print(analysis)
#   s <- analysis_samples[[analysis]]
#   
#   # design
#   on_icb_response <- s$time == "t1" & s$response == "yes"
#   on_icb_no_response <- s$time == "t1" & s$response == "no"
#   design <- model.matrix(~patient, data=s)
#   design_response <- cbind(design, on_icb_no_response, on_icb_response)
# 
#   y <- log2(g_qn_cpm + log2_prior_count)
#   y <- get_filtered_subset(norm_expr_mat[, s$library], 
#                            min.expr = de_min_log2_cpm, 
#                            min.prop = de_min_prop)
#   print(nrow(y))
#   res <- run_limma_trend(y, design_response, "on_icb_response", 
#                          de_padj_cutoff, de_log2fc_cutoff,
#                          analysis, method)
#   de <- bind_rows(de, res)
# }
# table(de$analysis, de$de)



table(fds$samples$path_aoi, fds$samples$slide)

s <- fds$samples %>% filter(path_specimen == "inv", segment == "panck")
s$path <- factor(tolower(s$path_aoi), levels=c("nl", "lgd", "hgd", "inv"))
y <- log2(normcounts)[,s$aoi]

# de analysis
de <- NULL
method <- "limma_trend"

design <- stats::model.matrix(~0 + slide + path, data=s)
colnames(design)

table(s$slide, s$path)




colnames(design) <- levels(s$path)
s$path
contrasts <- limma::makeContrasts(lgd_vs_nl = panck_lgd - panck_nmld,
                                  hgd_vs_nl = panck_hgd - panck_nmld,
                                  inv_vs_nl = panck_inv - panck_nmld,
                                  hgd_vs_lgd = panck_hgd - panck_lgd,
                                  inv_vs_lgd = panck_inv - panck_lgd,
                                  inv_vs_hgd = panck_inv - panck_hgd,
                                  invhgd_vs_lgdnl = (panck_inv + panck_hgd)/2 - (panck_lgd + panck_nmld)/2,
                                  levels=design)
fit <- run_limma_trend(y, design, contrasts)
res <- process_limma(fit, contrasts, method, de_padj_cutoff, de_log2fc_cutoff) 
de <- bind_rows(de, res)


model.matrix(~slide, data=s)

# 
# # design
# on_icb_response <- s$time == "t1" & s$response == "yes"
# on_icb_no_response <- s$time == "t1" & s$response == "no"
# design <- model.matrix(~patient, data=s)
# design_response <- cbind(design, on_icb_no_response, on_icb_response)
# 
# y <- log2(g_qn_cpm + log2_prior_count)
# y <- get_filtered_subset(norm_expr_mat[, s$library], 
#                          min.expr = de_min_log2_cpm, 
#                          min.prop = de_min_prop)
# print(nrow(y))
# res <- run_limma_trend(y, design_response, "on_icb_response", 
#                        de_padj_cutoff, de_log2fc_cutoff,
#                        analysis, method)
# de <- bind_rows(de, res)

write_xlsx(de, file.path(working_dir, "de_path.xlsx"))


res %>% filter(gene == "LAMB3")
res %>% filter(analysis == "hgd_vs_lgd") %>% arrange(padj)
table(res$analysis, (res$de == "up") & (res$padj < 1e-3))


write_xlsx(de)

plot_de_volcano(res) +
  facet_wrap(~ analysis)
  



```


## bg correction testing

```{r bg correction testing}

# params for plot
quantiles=c(0.10, 0.25, 0.5, 0.75, 0.90, 0.99)
width <- 16
height <- 6
  
# params for bgcorrect
bw.adjust <- 2
bg.quant <- 0.5

bg <- ds$meta$bg
x <- ds$counts


## plots

# raw cpm
x <- as_tibble(cpm(ds$counts))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_none_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_none_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, xraw)
# f <- file.path(plot_dir, "qc_bgcorrect_none_boxplot.pdf")
# ggsave(f, p, width=width, height=8)

# qq cpm
x <- st_geomx_bgcorrect(ds, method="qq", bw.adjust=bw.adjust, bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_qq_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_qq_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_qq_boxplot.pdf")
# ggsave(f, p, width=16, height=8)

# ppv cpm
x <- st_geomx_bgcorrect(ds, method="kdeppv", bw.adjust=bw.adjust, bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_ppv_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_ppv_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_ppv_boxplot.pdf")
# ggsave(f, p, width=16, height=8)

# bgsub cpm
x <- st_geomx_bgcorrect(ds, method="bgsub", bg.quant=bg.quant)
x <- as_tibble(cpm(x))
p <- st_geomx_plot_bg_dotplot(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_bgsub_dotplot.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)
p <- st_geomx_plot_bg_scatter(ds, x, quantiles)
f <- file.path(plot_dir, "qc_bgcorrect_bgsub_scatter.pdf")
ggsave(f, plot=p, device="pdf", width=width, height=height)

# p <- st_geomx_plot_dist_boxplot(ds, x)
# f <- file.path(plot_dir, "qc_bgcorrect_bgsub_boxplot.pdf")
# ggsave(f, p, width=16, height=8)



```


## BG correction version 4

```{r bgcorrect v4}

# parameters
eps <- 1e-10
n <- 2^13
bw.adjust <- 2
bg.quant <- 0.0

# choose sample
sort(ds$samples$bg_auc)[1:20]
sort(ds$samples$bg_auc, decreasing=TRUE)[1:10]

s <- c(ds$samples %>% slice_max(bg_auc, n=10) %>% pull(aoi), 
       ds$samples %>% slice_min(bg_auc, n=10) %>% pull(aoi))

for (i in s) {
  print(i)

  # setup
  bg <- ds$meta$bg
  x <- log2(unlist(ds$counts[,i]))
  xmin <- min(x)
  xmax <- max(x)
  
  expressed <- !bg & (x > quantile(x[bg], bg.quant))
  table(bg, expressed)
  
  # select bandwidth
  bwgene <- bw.nrd0(x[expressed]) * bw.adjust
  bwbg <- bw.nrd0(x[bg]) * bw.adjust
  bw <- max(bwgene, bwbg)
  
  d <- density(x[expressed], bw=bw, from=xmin, to=xmax, n=n)
  dcdf <- cumsum(d$y) / sum(d$y)
  
  dbg <- density(x[bg], bw=bw, from=xmin, to=xmax, n=n)
  dbgcdf <- cumsum(dbg$y) / sum(dbg$y)
  
  # false omission rate (FOR)
  dp <- dcdf / (dbgcdf + dcdf + eps)
  dp <- (dp - min(dp)) / (max(dp) - min(dp))
  
  # ensure isotonic behavior
  dp.ir <- isoreg(d$x, dp)$yf
  
  # interpolate
  p <- approx(d$x, dp.ir, xout=x, ties="ordered")$y
  #f <- splinefun(d$x, dp.ir, method="natural", ties="ordered")
  #p <- f(x)
  
  # transform
  y <- (2^x) * p
  y <- pmax(1, y)
  
  tbl <- bind_cols(dx = d$x,
                   dcdf = dcdf,
                   dbgcdf = dbgcdf,
                   dfor = dfor,
                   dp = dp,
                   dp.ir = dp.ir)
  
  ggplot(tbl) +
    geom_line(aes(dx, dcdf)) +
    geom_line(aes(dx, dbgcdf), color="red") +
    geom_line(aes(dx, dp), color="blue", alpha=0.8) +
    geom_line(aes(dx, dp.ir), color="blue", linetype="dashed", alpha=0.8) +
    geom_vline(xintercept=median(x[bg]), color="red", linetype="dashed", alpha=0.5)
  
  
  tbl <- bind_cols(x = 2^x, 
                   y = y)
  print(ggplot(tbl) +
    geom_line(aes(x, x), color="black", alpha=0.5) +
    geom_line(aes(x, y), color="red", alpha=0.5) + 
    scale_x_log10() +
    scale_y_log10())

}



# choose samples
sort(ds$samples$bg_auc)[1:20]
sort(ds$samples$bg_auc, decreasing=TRUE)[1:10]
s <- c(ds$samples %>% slice_max(bg_auc, n=10) %>% pull(aoi), 
       ds$samples %>% slice_min(bg_auc, n=10) %>% pull(aoi))

bg <- ds$meta$bg

for (i in s) {
  print(i)
  x <- unlist(ds$counts[,i])
  #res <- bgcorrect_kdeppv(x, bg, bw.adjust=2, bg.quant=0.5)
  #res <- bgcorrect_kdefor(x, bg, bw.adjust=2, bg.quant=0.5)

  tbl <- bind_cols(dx = res$dx,
                   dcdf = res$dcdf,
                   dbgcdf = res$dbgcdf,
                   dp = res$dp)

  print(ggplot(tbl) +
    geom_line(aes(dx, dcdf)) +
    geom_line(aes(dx, dbgcdf), color="red") +
    geom_line(aes(dx, dp), color="blue", alpha=0.8) +
    geom_vline(xintercept=median(log2(x)[bg]), color="red", linetype="dashed", alpha=0.5)
  )
  
  tbl <- bind_cols(x = x, 
                   y = res$y)
  print(ggplot(tbl) +
          geom_line(aes(x, x), color="black", alpha=0.5) +
          geom_line(aes(x, y), color="red", alpha=0.5) + 
          scale_x_log10() +
          scale_y_log10()
  )
  
}


```

## BG correction exploration

```{r bg correction exploration}

bg <- ds$meta$bg
x <- ds$counts

sort(ds$samples$bg_auc)[1:20]
sort(ds$samples$bg_auc, decreasing=TRUE)[1:10]
s <- "s08_r010_panck"
xs <- unlist(x[,s])

bw.adjust <- 5

x <- log2(xs)
xmin <- min(x)
xmax <- max(x)

n <- 10000
bw <- bw.nrd0(x) * bw.adjust
eps <- 1e-10

dbg <- density(x[bg], bw=bw, from=xmin, to=xmax, n=n)

?quantile


#d <- density(x[!noise], bw=bw, from=xmin, to=xmax, n=n)

dcdf <- cumsum(d$y) / sum(d$y)
dbgcdf <- cumsum(dbg$y) / sum(dbg$y)
dbgcdf <- pmax(dbgcdf, dcdf)

dfdr <- (1-dbgcdf) / (1-dcdf+eps)
dfdr <- pmin(1, dfdr + fdr.min)
scaled_fdr <- xmin + (xmax - xmin) * cumsum(dfdr) / n
scaled_fdr <- pmin(d$x, scaled_fdr)
xnoise <- approx(d$x, scaled_fdr, xout=x)$y
xnoise <- pmin(xnoise, x)
y <- 2^x - 2^xnoise + 1

#dfdr <- (1-dbgcdf) / (1-dcdf)
#dfdr[which(is.nan(dfdr))] <- min(dfdr[which(!is.nan(dfdr))])
#dfdr <- pmin(1, dfdr + fdr.min)



res <- cluster_noise(xs, bg)

tbl <- bind_cols(x=res$x,
                 bg=res$bg_dist,
                 hi=res$hi_dist,
                 tot=res$bg_dist + res$hi_dist)

ggplot(tbl) +
  geom_point(aes(x,bg), color="red", alpha=0.5) +
  geom_point(aes(x,hi), color="blue", alpha=0.5) +
  geom_point(aes(x,tot)) +
  geom_vline(xintercept=res$bg_center, linetype="dashed", color="red") +
  geom_vline(xintercept=res$hi_center, linetype="dashed", color="blue") +
  geom_vline(xintercept=res$cutoff, linetype="dashed")


res <- bgcorrect_kdefdr2(xs, bg, bw.adjust=1, fdr.min=0.05)$y


for (s in ds$samples$aoi) {
  print(s)
  xs <- unlist(x[,s])
  res <- bgcorrect_kdefdr2(xs, bg, bw.adjust=1, fdr.min=0.05)$y
}



```

## DE exploration

```{r de lme4}
library(lmerTest)


# bgcorrect_method <- "bgsub"
# norm_method <- "qn"
# prefix <- paste0(bgcorrect_method, "_", norm_method)
# 
# x <- st_geomx_bgcorrect(fds, method=bgcorrect_method,
#                         bw.adjust=bgcorrect_bw_adjust,
#                         bg.quant=bgcorrect_bg_quant)
# x <- st_geomx_normalize(fds, x, method=norm_method)
# 
# 
# # design
# s <- filter(fds$samples, path_specimen == "inv")
# s$patient <- factor(s$patient)
# s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
# 
# design <- model.matrix(~0 + path_aoi + patient, s)
# 
# colnames(design)
# 
# contrasts <- makeContrasts(inv_vs_nl = path_aoiinv - path_aoinl,
#                            hgd_vs_nl = path_aoihgd - path_aoinl,
#                            lgd_vs_nl = path_aoilgd - path_aoinl,
#                            levels=design)
# 
# y <- log2(x[,s$aoi])
# res <- run_limma_trend(y, design, contrasts, 
#                        padj_cutoff=de_padj_cutoff, 
#                        log2fc_cutoff=de_log2fc_cutoff,
#                        analysis=prefix)
# 
# res %>% arrange(desc(log2fc))
#
#
# run_limma_trend <- function(y, design, coef, padj_cutoff, log2fc_cutoff, 
#                             analysis, method) {
#   fit <- lmFit(y, design)
#   fit <- eBayes(fit, trend=TRUE)
#   res <- topTable(fit, coef=coef, number=Inf, sort.by="none")
#   res <- limma_rename_result_cols(res)
#   res <- de_format_results(res, analysis=analysis, method=method,
#                            padj_cutoff = padj_cutoff, log2fc_cutoff = log2fc_cutoff)
#   return(res)
# }

bgcorrect_method <- "bgsub"
bgcorrect_bw_adjust <- 2
bgcorrect_bg_quant <- 0.5
norm_method <- "rle"

x <- st_geomx_bgcorrect(fds, method=bgcorrect_method,
                        bw.adjust=bgcorrect_bw_adjust,
                        bg.quant=bgcorrect_bg_quant)
x <- st_geomx_normalize(fds, x, method=norm_method)


# setup for de analysis
s <- filter(fds$samples, path_specimen == "inv")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
y <- log2(x[,s$aoi])

g <- "KRT17"

tbl <- bind_cols(select(s, patient, path_aoi), y=y[g,])

fit <- lmerTest::lmer(y ~ path_aoi + (1|patient), data=tbl)
summary(fit)
ls_means(fit)

# Now check the anova-output:
lmer1 <- lmer(sweet~product+(1|assessor), data=sensintro)
anova(lmer1)


# Analysing all 20 assessors:
lmer1 <- lmer(sweet~product+(1|assessor), data=sensintro)
summary(lmer1)

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(inv_vs_nl = path_aoiinv - path_aoinl,
                           hgd_vs_nl = path_aoihgd - path_aoinl,
                           lgd_vs_nl = path_aoilgd - path_aoinl,
                           levels=design)

```
